## Introduction

<!-- Kinase signaling / phophoproteomics in general -->

Cell signaling networks formed by protein kinases regulate cell fate and behavior through protein phosphorylation [@PMID:7834742]. As such, it is not surprising that kinase dysregulation orchestrates the onset and development of a myriad of diseases, especially in cancer. Despite the overwhelming focus on genomics, the analysis of signaling behaviors involved in cancer measured by mass spectrometry (MS)-based global phosphoproteomics provides a great opportunity to direct translational opportunities [@PMID:30670634]. Nevertheless, despite the steep accumulation of large-scale phosphoproteomic clinical data, one of the biggest challenges remains the systematic identification of signaling effectors driving cancer cells.

<!-- Current methods to analyze phosphoproteomics -->

One approach to make sense of phosphoproteomic measurements has been to infer the activity of upstream kinases. Previously published methods combine the phosphorylation signal with reported kinase-substrate interactions to reconstruct signaling networks. For instance, kinase-substrate enrichment analysis (KSEA) averages the signals of groups of kinase substrates to infer enriched pathways in biological samples [@PMID:23532336]. Another method, Integrative Inferred Kinase Activity (INKA), infers kinase activity by integrating the scores of two kinase-centric components; overall kinase phosphorylation levels and activation loop phosphorylation, and two substrate-centric components; quantification of experimentally reported kinase-substrate relationships shown in PhosphoSitePlus and levels of kinase substrates predicted by an algorithm–NetworKIN–that performs motif-based predictions to pinpoint kinases responsible for specific phosphorylation events [@PMID:31126969; @PMID:30445427; @PMID:17570479]. Using kinase specificity profiles generated from oriented peptide library scanning experiments, Scansite predicts kinase-substrate interactions while identifying the sequence motifs that are recognized by specific modular signaling protein domains [@PMID:12824383]. 

<!-- Limitations of these methods; (1) Kinase prediction bias (2) Data missingness -->

These methods, sometimes in combination, help reconstruct signaling networks from highly complex coordinate phosphoproteomic measurements to find those signaling pathways and molecules driving phenotypic states. However, kinase prediction methods are necessarily dependent on characterized kinase-substrate interactions and unfortunately the majority of the phosphoproteome remains largely uncharacterized [@PMID:30670635]. In fact, just 20% of kinases have been shown to phosphorylate 87% of currently annotated substrates and around 80% of kinases have fewer than 20 substrates, and 30% are yet to be assigned a single substrate [@PMID:30670635]. Hence, insights generated by computational methods dependent on this unequal distribution of known interactions are less likely to predict understudied protein kinases to be causally involved. Importantly, an additional major challenge being faced during the analysis of large-scale signaling data is missingness. Given that multiplexed tandem mass tag (TMT) MS experiments process batches of a limited number of samples at a time, the signaling coverage among samples is different [@PMID:19921851]. This means that peptides that are measured in a particular batch of 10 specimens may not be detected in other samples. Computational tools usually require complete data sets which is why a frequent strategy to handle this challenge is either imputing missing values with a representative statistic (e.g. average signal) or throwing any peptides displaying missing values––at the expense of losing critical information. Hereby, the power of kinase enrichment and prediction methods may be further compromised by this problem. Thus, there is a clear need to develop tailored unbiased computational methods capable of modeling the entirety of the phosphoproteomic data set––including its missingness––and we propose data-dependent modeling as an alternative route to do so.

<!-- Introduction to paper -->

Here, we present an EM algorithm that probabilistically and simultaneously models both the peptide phosphorylation levels with a Gaussian Mixture Model (GMM) and the peptide sequences using either a PAM250 or binomial position enrichment model (Fig. 1) to reconstitute cell signaling networks and identify causal interactions. Widely used clustering methods such as hierarchical clustering or k-means intent to disentangle signaling nodes by grouping phosphorylation site co-variation. However, while this clustering criterion alone results in groups of peptides that display similar activation patterns across conditions, they may be regulated by sets of different upstream kinases driving different phenotypic consequences. To overcome this limitation, our clustering algorithm assigns phosphopeptides based on phosphorylation status but also sequence similarity. Studies have shown that the residues surrounding phosphorylation sites have had to evolve throughout millions of years to become exquisitely fine-tuned motifs that confer signaling specificity and fidelity [@PMID:14668868; @PMID:19589966]. However, clustering based on motif similarity alone could result in groups of proteins that while sharing the same set of upstream kinases, their individual regulation may be context-dependent. In this study, we show that careful interpretation of the identified clusters facilitates the identification of upstream regulating kinases experimentally validated in the literature to either strongly favor an enriched sequence motif or to interact with particular proteins within a cluster. To test the utility of our method, we analyze the phosphoproteomes of 110 treatment-naïve lung adenocarcinoma (LUAD) tumors and 101 paired normal adjacent tissues (NATs) from the National Cancer Institute (NCI)’s Clinical Proteomic Tumor Analysis Consortium (CPTAC) LUAD study [@PMID:32649874] in order to systematically identify groups of signaling molecules that coordinately drive molecular and clinical features.
