## Results


### Constructing an expectation-maximization algorithm tailored to cluster phosphoproteomic data

![**Schematic of the DDMC approach to cluster global signaling data and infer upstream kinases driving phenotypes.** A) DDMC simultaneously models phosphorylation signal and peptide sequence to cluster global signaling measurements. B) Modeling based on clusters such as regression methods allows establisheing assocations between specific clusters and features of interest. C) Putative upstream kinases regulating meaningful clusters can be predicted by computing the distance between a cluster motif's PSSM and a PSPL PSSM. PSSM; Position-specific scoring matrix, PSPL; Position scanning peptide library](DDMC_functional_diagram.svg "Figure 1"){#fig:method width="100%"}

MS-based global phosphoproteomic data provides unparalleled coverage when interrogating kinase signaling networks and their therapeutic implications. However, these data also present challenging issues as a consequence of their incomplete and stochastic coverage, high-content but low-sample throughput, and variation in coverage across experiments. In addressing these issues, we recognized that MS measurements simultaneously provide two pieces of information—one obtains the exact site of phosphorylation on a peptide sequence and then some measure of abundance within the sample of interest. Both of these pieces of information are critical to the overall interpretation of the data.

Based on this observation, we built a mixture model that probabilistically clusters phosphosites based on both their peptide sequence and abundance across samples (Fig. {@fig:methodS}). In each iteration, DDMC applies an expectation-maximization algorithm to optimize clusters that capture the average response of member peptides and a peptide motif, in the form of a position-specific scoring matrix (PSSM), summarizing the sequence features of that cluster (Fig. {@fig:method}A and {@fig:methodS}). Clustering both the sequence and abundance measurements simultaneously ensures that the resulting clusters are a function of both types of features, which we hypothesized would provide both more meaningful and robust clusters.

The resulting clustering provides coordinated output that can be used in a few different ways. The cluster centers, by virtue of being a summary for the abundance changes of these peptides, can be regressed against phenotypic responses (e.g., cell phenotypes or clinical outcomes) to establish associations between particular clusters and response (Fig. {@fig:method}B). This can help to identify which cluster of peptides drives a certain response. Regression using the clusters instead of each peptide ensures that the model can be developed despite relatively few samples, with minimal loss of information since each peptide varies in a similar manner.

In parallel or independently, one can interrogate the resulting PSSMs to describe the overall sequence features of that cluster. These outputs can be readily compared to other information such as experimentally generated specificity profiles of putative upstream kinases via Position Specific Scanning Libraries (PSPL), to infer upstream kinases [@PMID:10945990; @PMID:20159853; @PMID:31124786; @PMID:15782149; @PMID:26551075]. We extracted a collection of 62 kinase specificity profiles to find what cluster motifs resemble most accurately the optimal motif of putative upstream kinases (Fig. {@fig:method}C) [@PMID:18765831; @PMID:15782149; @PMID:26551075]. However, kinase-substrate specificity is also dictated by features outside of the immediate substrate region, and we note that our approach is more general than strictly assembling kinase-substrate predictions. Overall, this overview demonstrates how DDMC can take complex, coordinated signaling measurements and find patterns in the phosphorylation signals to reconstruct signaling networks and associate particular clusters and phenotypes.

### Dual data-motif clustering strategy robustly imputes missing values

![**Benchmarking the robustness of motif clustering to missing measurements.** A) A schematic of the process for quantifying robustness to missing values. Any peptides containing less than 7 TMT experiments were discarded. For the remaining 15904 peptides, an entire random TMT experiment was removed per peptide and these values were stored for later comparison. Next, these artifical missing values were imputed using either a baseline strategy (peptide mean/minimum signal) or the corresponding cluster center. Once a mean squared error was computed for each peptide, the second iteration repeats this process by removing a second TMT experiment. A total of 5 random TMT experiments per peptide were imputed by clustering using either the peptides' phosphorylation signal only (B), the peptides' sequence only (C) or a mix of both information sources (D). The imputation errors were calculated across different number of clusters (E) and different data-vs-sequence weights (F).](figureM2.svg "Figure 2"){#fig:missingness width="100%"}

To evaluate the robustness of our combinad dual data-motif clustering (DDMC) method in analyzing incomplete data sets, we designed a computational experiment wherein we removed specific observations and attempted to predict them using the cluster centers corresponding to the peptides those missing values belonged to (Figure 2A). The resulting mean squared errors between the actual and predicted values were compared to compared to commonly used imputation strategies such as the peptides' mean or minimum signal. Furthermore, we evaluated the imputation performance of our method when clustering the data by phosphorylation signal only, by sequence similarity only, or by combining both information sources (2B-D). These results indicate that MDCC clearly imputes missing values more accurately than the peptides' mean signal in all three cases. No noticeable differences were observed in the imputations performance using data or sequence only, or a mix of both. Next, we re-implemented this strategy to find out what number of clusters (2E) and weight (2F) yielded more accurate predictions. We found that a larger number of clusters and smaller weights (prioritizing phosphorylation signal over peptide sequence) minimized the imputation error. These data suggests that DDMC outperforms standard imputation strategies such as peptide mean signal (and minimum signal).   

#TODO: 
- Compare to minimum signal
- Filling the missingness with 0's
- Matrix completion by PCA
- Add legend for each regression line in plots B-F and mix labels/titles


### Shared differences between tumor and normal adjascent tissue

![**Conserved tumor differences compared to normal adjascent tissue.** A) Hierarchical clustering of DDMC cluster centers. B–C) Principal components analysis scores (B) and loadings (C) of the samples and phosphopeptide clusters, respectively. D) Phosphorylation signal of tumor and NAT samples per cluster and statistical significance according to a Mann Whitney rank test. E) lasso receiver operating characteristic curve. F) lasso weights per cluster. G-I) Logo plot for the motifs of clusters 11 (G), 12 (H), and 23 (I). (J) Heatmap of Frobenius norms between the PSSMs of clusters 11, 12, and 23 and the PSPL matrices of putative upstream kinases.](figureM3.svg "Figure 3"){#fig:TvsNAT width="100%"}

<!-- Intro to model set up and data -->
We explored whether DDMC could recognize conserved signaling patterns in tumors compared to normal adjacent tissue (NAT). To do so, we utilized DDMC with the binomial sequence distance method and 24 clusters (Fig. {@fig:method}, {@fig:missingness}E/F). We were able to include many peptides that were not observed in every tumor through our ability to handle missing data but filtered out any peptides that were not captured in at least two 10-plex TMT runs, resulting in a data set of 30,561 peptides. We used this fitting result throughout the rest of this study. The resulting 24 cluster motifs can be found in Fig. {@fig:motifsS}.

<!-- The signaling in both tissues is so different that any modeling strategy captures distinct patterns -->
The signaling difference between tumors and NAT samples was substantial, highlighting the significant signaling rewiring that tumor cells must undergo (Fig. {@fig:TvsNAT}A). Using principal components analysis, we could observe that NAT samples were more similar to one another than each tumor sample (Fig. {@fig:TvsNAT}B/C). Every cluster was significantly different in its average abundance between tumor and NAT (Fig. {@fig:TvsNAT}B/C). Not surprisingly given these enormous differences, samples could be almost perfectly classified using their phosphopeptide signatures, with or without DDMC (Fig. {@fig:TvsNAT}E; {@fig:TvsNATS}).

<!-- DDMC data -->
Hierarchical clustering of the DDMC cluster centers shows three distinct groups of clusters. One group had lower tumor abundances as compared to NATs (e.g., 11, 7, 19), another group showed the opposite (e.g., 1, 2, 12), and a few additional clusters either varied randomly (4, 21, 13) or showed inconsistent variation with tumor versus NAT status (9, 17, 20) (Fig. {@fig:TvsNAT}A). Consistent with the hierarchical clustering, the PCA loading of cluster 11 is negatively weighted along PC1, following all NAT samples (Fig. {@fig:TvsNAT}B). Conversely, clusters 1 and 12, among others, were positively weighted on PC 1 (Fig. {@fig:TvsNAT}C).

<!-- DDMC sequence -->
With the abundance changes we observed, we decided to further explore clusters 11, 12, and 23. Cluster 11 displays a strong conservation for lysine across all peptide positions, a clear enrichment of phenylalanine at -1, and an enrichment of tyrosine, lysine, methionine and histidine at +1 (Fig. {@fig:TvsNAT}G). In contrast, cluster 12 is enriched for acidic residues in most positions along with enrichment of tyrosine at -1, aspartic acid and tyrosine at +1, and lysine at +2 (Fig. {@fig:TvsNAT}H). The presence of tyrosines flanking the phosphorylation site of clusters 11 and 12 might suggest the activity of a coordinated regulation of this cluster wherein upon an initial phosphorylation event at either -1 or +1, the substrate may become activated for subsequent phosphorylation at the central serine [@PMID:26551075]. The sequence motif of cluster 23 is characterized by an overrepresentation of phenylalanine across several positions, especially at -3. By comparing the cluster PSSMs with our collection of specificity profiles of putative upstream kinases, we found that cluster 11 might be modulated by SLK and/or PAK2; cluster 12 by CK2, and cluster 23 by ERK2 (Fig. {@fig:TvsNAT}J). Taken together, DDMC builds phosphoproteomic clusters that present signaling dysregulation common to tumors compared to NATs and identifies putative upstream kinases modulating them. These features can help to interpret phosphoproteomic results, even when there are substantial differences between samples.

<!-- TODO: THE FACT THAT CLUSTERS 11 AND 23 ARE UPREGULATED IN NATs WOULD SUGGEST THAT SLK, PAK2, AND ERK2 ACT AS TUMOR SUPPRESSORS... -->

### Phosphoproteomic aberrations associated with genetic tumor differences

![**Phosphoproteomic aberrations associated with STK11 mutational status.** A) Phosphorylation signal of STK11 WT and mutant samples per cluster and statistical significance according to a Mann Whitney rank test B) ROC of a lasso model predicting the STK11 mutational status and (C) its corresponding weights per sample type (D-G) Logo plots of PSSM motifs of clusters 1, 8, 9, and 15. H) Heatmap of putative upstream kinases.](figureM4.svg "Figure 4"){#fig:STK11 width="100%"}

<!-- Brief intro to STK11 -->
STK11 in LUAD.... 
Thus, we aimed to identify the phosphoproteomic aberrations triggered by the mutational status of STK11.

<!-- STK11 DDMC data  -->
The vast majority of clusters were statistically significant and overall they had higher mutant abundances as compared to WT. Among others, clusters 1, 5, 12, 15, 21, 22, and 24 displayed an upregulation of phosphorylation signal in STK11 mutant samples versus WT whereas clusters 11, 14, and 19 showed the opposite trend (Fig. {@fig:STK11}A). We then utilized the cluster centers corresponding to each patient's tumor and NAT samples to succesfully predict the STK11 mutational status using a lasso model. The underlying signaling differences are mainly explained by the tumor samples of clusters 8 and 14 to which the model assigned very large negative weights(Fig. {@fig:STK11}B). These results motivated further exploration of clusters 8, 14, and 22 .

<!-- STK11 DDMC sequence -->
The motif of cluster 8 is characterized by a strong enrichment of hydrophobic residues in several positions such as tryptophan at +1 and +2, as well as leucine, isoluecine, phenylalanine, and methionine at -5, +4 and +5. In addition, this motif might also be jointly regulated by serine kinases at positions -1 and -2. Cluster 14 shows a PSSM motif with a clear enrichment of basic residues in its N-terminus, especially at positions -3 and -2. Also, hydrophobic residues seem to be favored at positions -4, -1 and +1, +4, and +5. The motif of cluster 22 displays a striking enrichment of all three basic residues at positions +1, +2, and +3, an overrepresentation of tyrosines at positions -5, -4, and to a lesser extent -2, and a strong presence of phenyalalanine at -2. (Fig. {@fig:STK11}D-F). By comparing the PSSMs of these motifs to the specificity profiles of putative upstream kinases, we found that cluster 1 sequence closely resembles the optimal target motif of AMPK; cluster 8 might be modulated by several NEK kinases 4, 5, 6 and 8 as well as PKC-ζ; cluster 14 displays a motif clearly favored by PKA and PKD, and to a lesser extent AKT; lastly, ERK1, NEK9, and SHC1 might be good candidates for the peptide motif of cluster 22 (Fig. {@fig:STK11}G). 

<!-- EGFRm/ALKf intro -->
Tyrosine kinase inihibitors (TKIs) targeting the receptor tyrosine kinases (RTKs) EGFR and ALK are effective treatments in cancer patients with EGFR mutations and/or ALK translocations. However, these treatments are limited by drug resistance which in some cases can be mediated by the concomitant signaling of both RTKs activated by driver mutations [@PMID:21791641; @PMID:27707887]. Therefore, we explored the conserved tumor signaling changes of patients harboring mutations in EGFR and/or gene fusions in ALK (EGFRm/ALKf).

<!-- EGFRm/ALKf data -->
Except three clusters, all of them showed high statistical significance with a marked upregulation of phosphorylation signal in EGFRm/ALKf samples compared with their WT counterparts. Clusters with the highest signal increase in mutant samples include 1, 2, 8, and 12 (Fig. {@fig:EGFRm/ALKf}A). Moreover, a lasso model predicted clusters 1, 13, 20, and 23 to mostly explain these signaling differences (Fig. {@fig:EGFRm/ALKf}B&C). 

<!-- EGFRm/ALKf sequence -->
The sequence motif of cluster 1 is enriched with basic residues at most positions, especially in the case of lysine at positions +2, +3, and +4. In addition, the presence of tyorsine at +1 in this motif could suggest that several peptides in this cluster may be coordinately primed by tyrosine kinases prior or after its central phosphorylation. There is also an overrepresentation of peptides displaying a cysteine in multiple positions, particularly at -3 and -2. Clusters 13 and 23 are clearly enriched with polar and hydrophobic residues, which as mentioned above, is a preferred motif for the pro-mitotic kinase ERK2. Cluster 13 might be modulated by NEK3 and DAPK3. Finally, cluster 20 shows an overall enrichment of acidic and hydrophobic residues, with a significant overrepresentation of isoleucine at several positions which does not seem to provide an ideal docking environment for any of the candiate upstream kinases. Our model identifies AMPK and p38 as the upstream kinases for clusters 1 and 13, respectively (Fig. {@fig:EGFRm/ALKf}D-F and Fig {@fig:TvsNAT}I).

<!-- Comparisong to other regression strategies -->
Finally, we compared the classification performance of four lasso models fit to either the DDMC clusters, clusters generated by the standard methods GMM and k-means, or the raw phosphoproteomic data directly. It is worth noting that—unlike DDMC—GMM, k-means, and direct regression can not handle missing values and thus for this strategies we used the 1,311 peptides that were observed in all samples, whereas DDMC was fit to the entire data set comprising 30,561 phosphosites. Interestingly, we found that samples were classified with higher accuracy with DDMC than using GMM and similarly than with k-means, especially with STK11 (Fig {@fig:RegressionsS}A). Direct regression to the raw signaling data yielded excellent performance, however this strategy assigns thousands of coefficients to different peptides that vary every time the model is run which renders this appraoch useless (Fig {@fig:RegressionsS}).  

<!-- TODO: AS IN FIGURE M3, DISCUSS DIRECTIONALITY, HERE WITH THE ADDED DIFFICULTY OF EACH CLUSTER HAVING A NAT AND A TUMOR COEFFICIENT -->

### Phosphoproteomic aberrations associated with clinical features

![**Figure 5.** Its caption.](figureM5.svg "Figure 5"){#fig:clusters width="100%"}

 #TODO in figures 3, 4, and 5:
- Benchmark upstream kinase predictions with:
    -> KSEA
    -> KinomeXplorer
    -> Scansite
