## Results


### Constructing an expectation-maximization algorithm tailored to cluster phosphoproteomic data

![**Schematic of the DDMC approach to cluster global signaling data and infer upstream kinases driving phenotypes.** A) DDMC simultaneously models phosphorylation signal and peptide sequence to cluster global signaling measurements. B) Modeling based on clusters such as regression methods allows establisheing assocations between specific clusters and features of interest. C) Putative upstream kinases regulating meaningful clusters can be predicted by computing the distance between a cluster motif's PSSM and a PSPL PSSM. PSSM; Position-specific scoring matrix, PSPL; Position scanning peptide library](EM_diagram.svg "Figure 1"){#fig:method width="100%"}

MS-based global phosphoproteomic data provides unparalleled coverage when interrogating kinase signaling networks and their therapeutic implications. However, these data also present challenging issues as a consequence of their incomplete and stochastic coverage, high-content but low-sample throughput, and variation in coverage across experiments. In addressing these issues, we recognized that MS measurements simultaneously provide two pieces of information—one obtains the exact site of phosphorylation on a peptide sequence and then some measure of abundance within the sample of interest. Both of these pieces of information are critical to the overall interpretation of the data.

Based on this observation, we built a mixture model that probabilistically clusters phosphosites based on both their peptide sequence and abundance across samples (Supplemental Fig. 1). In each iteration, DDMC applies an expectation-maximization algorithm to optimize clusters that capture the average response of member peptides and a peptide motif, in the form of a position-specific scoring matrix (PSSM), summarizing the sequence features of that cluster (Fig. {fig:method}A and Supplemental Fig. 1). Clustering both the sequence and abundance measurements simultaneously ensures that the resulting clusters are a function of both types of features, which we hypothesized would provide both more meaningful and robust clusters.

The resulting clustering provides coordinated output that can be used in a few different ways. The cluster centers, by virtue of being a summary for the abundance changes of these peptides, can be regressed against phenotypic responses (e.g., cell phenotypes or clinical outcomes) to establish associations between particular clusters and response (Fig. {fig:method}B). This can help to identify which cluster of peptides drives a certain response. Regression using the clusters instead of each peptide ensures that the model can be developed despite relatively few samples, with minimal loss of information since each peptide varies in a similar manner.

In parallel or independently, one can interrogate the resulting PSSMs to describe the overall sequence features of that cluster. These outputs can be readily compared to other information such as experimentally generated specificity profiles of putative upstream kinases via Position Specific Scanning Libraries (PSPL), to infer upstream kinases [@PMID:10945990; @PMID:20159853; @PMID:31124786; @PMID:15782149; @PMID:26551075]. We extracted a collection of 62 kinase specificity profiles to find what cluster motifs resemble most accurately the optimal motif of putative upstream kinases (Fig. {fig:method}C) [@PMID:18765831; @PMID:15782149; @PMID:26551075]. However, kinase-substrate specificity is also dictated by features outside of the immediate substrate region, and we note that our approach is more general than strictly assembling kinase-substrate predictions. Overall, this overview demonstrates how DDMC can take complex, coordinated signaling measurements and find patterns in the phosphorylation signals to reconstruct signaling networks and associate particular clusters and phenotypes.

### Dual data-motif clustering strategy robustly imputes missing values

To evaluate the robustness of our combinad dual data-motif clustering (DDMC) method in analyzing incomplete data sets, we designed a computational experiment wherein we removed specific observations and attempted to predict them using the cluster centers corresponding to the peptides those missing values belonged to (Figure 2A). The resulting mean squared errors between the actual and predicted values were compared to compared to commonly used imputation strategies such as the peptides' mean or minimum signal. Furthermore, we evaluated the imputation performance of our method when clustering the data by phosphorylation signal only, by sequence similarity only, or by combining both information sources (2B-D). These results indicate that MDCC clearly imputes missing values more accurately than the peptides' mean signal in all three cases. No noticeable differences were observed in the imputations performance using data or sequence only, or a mix of both. Next, we re-implemented this strategy to find out what number of clusters (2E) and weight (2F) yielded more accurate predictions. We found that a larger number of clusters and smaller weights (prioritizing phospho-signal over peptide sequence) minimized the imputation error. These data suggests that DDMC outperforms standard imputation strategies such as peptide mean signal (and minimum signal).   

#TODO: 
- Compare to minimum signal
- Filling the missingness with 0's
- Matrix completion by PCA
- Add legend for each regression line in plots B-F and mix labels/titles

![**Benchmarking the robustness of motif clustering to missing measurements.** A) A schematic of the process for quantifying robustness to missing values. Any peptides containing less than 7 TMT experiments were discarded. For the remaining 15904 peptides, an entire random TMT experiment was removed per peptide and these values were stored for later comparison. Next, these artifical missing values were imputed using either a baseline strategy (peptide mean/minimum signal) or the corresponding cluster center. Once a mean squared error was computed for each peptide, the second iteration repeats this process by removing a second TMT experiment. A total of 5 random TMT experiments per peptide were imputed by clustering using either the peptides' phosphorylation signal only (B), the peptides' sequence only (C) or a mix of both information sources (D). The imputation errors were calculated across different number of clusters (E) and different data-vs-sequence weights (F).](figureM2.svg "Figure 2"){#fig:tn width="100%"}


### Shared differences between tumor and normal adjascent tissue

We explored whether DDMC could recognize conserved signaling patterns in tumors compared with normal adjacen tissue (NAT) samples. To do so, we utilized DDMC with the binomial sequence distance method, setting the number of clusters to 24 and a sequence vs data weight of 15 (Fig. 1, 2E, F). We filtered any peptides out that were not captured in at least two 10-plex TMT experiments which rendered a final data set of 30.561 peptides. We used this fitting for all analyses performed in this study. Hierarchical clustering of the cluster centers shows a striking contrasting signal between tumor and NAT samples (Fig 3A). This heatmap shows two distinctly regulated groups of NAT samples (top and bottom) whereas all tumor samples cluster together in the middle. Three distinct signaling patterns can be observed; (i) high signal in tumors and low in both NAT groups, (ii) heterogeneous signal in tumors, a high NAT group (top) and a low NAT group (bottom), and (iii) cluster 11 shows a unique signal wherein these peptides are enriched in most NAT samples and depleted in all tumors. The scores of the PCA analysis of the cluster centers further underscore the signaling differences between tumors—positively weighted across principal component (PC) 1—and NAT samples—negatively weighted across PC1—(Fig. 3B). Consistent with the hierarchical clustering, the PCA loading of cluster 11 is negatively weighted along PC1, closely tracking with all NAT samples. Conversely, clusters 12 and 1, among others, display similar PC1 weights as tumor samples (Fig. 3C). Additionally, we asked whether the signaling between tumor versus NAT samples for every cluster were statistically different. We performed Mann Whitney rank tests and found that all clusters display significantly different phosphorylation levels in their tumor samples compared with NATs (Fig. 3C&D). As shown in the hierarchical clustering, the overall signal of tumor samples is dramatically lower than in NAT samples in clusters 11, 19 and 23, whereas clusters 1, 2, and 12 (to name only a few) show the opposite pattern. Next, we fit a lasso logistic regression model to the cluster centers to predict sample type (tumor or NAT). All samples were correctly predicted except from one, again highlighting the striking differences in signaling between cancer and normal cells (Fig. 3E). Most of these signaling differences were expleined by cluster 11 (Fig. 3F). With this information, we decided to further explore clusters 11, 12, and 23. By looking at what residues where enriched across all peptide positions, we found that cluster 11 displays a strong conservation for lysine throughout all positions—but especially in the C-terminus—, a clear enrichment of phenylalanine at -1, and an enrichment of tyrosine, lysine, methionine and histidine at +1 (Fig. 3G). In contrast, cluster 12 displays a motif with a clear enrichment for acidic residues in most positions as well as a noticeable enrichment of tyrosine at -1, aspartic acid and tyorsine at +1, and lysine at +2 (Fig. 3H). The presence of tyrosines flanking the phosphorylation site of clusters 11 and 12 might suggest the activity of a coordinated regulation of this cluster wherein upon an initial phosphorylation event at either -1 or +1, the substrate becomes activated for subsequent phosphorylation at the central serine [@PMID:26551075]. The sequence motif of cluster 23 is characterized by an overrepresentation of phenylalanine across several positions, especially at -3. By comparing the cluster PSSMs with our collection of specificity profiles of putative upstream kinases, we found that cluster 11 might be modulated by SLK and PAK2; cluster 12 by CK2, and cluster 23 by ERK2 (Fig. 3J). SLK has been shown to play a role in cancer cell motility and invasiveness via interaction with SRC&CK2 [@PMID:16837460] and HER2 [@PMID:19525980] in addition to cytoskeletal remodeling proteins. Moreover, a novel ALK inhibitor ensartinib (X-396) that has shown promising clinical activity in patients with ALK-positive NSCLC in a phase 1/2 trial [@PMID:29563138] target a handful of additional kinases including SLK. Nevertheless, the invovlement of SLK in LUAD has not been explored yet. Kikuchi et al observed an overexpression of PAK2 in non-small cell lung cancer (NSCLC) patient samples via a shotgun proteomics analysis which was subsequently validated by histoimmunochemistry [@PMID:22761400]. The inhibition of CK2 has been shown to sensitize NSCLC tumors to cisplatin via upregulation of PML [@PMID:28744813] and to inionizing radiation [@PMID:29170453]. Lastly, ERK2 is an frequently observed as a prognostic marker of resistance to targeted therapies inducing tumorigenesis in several cancer types. Brognard et al demonstrated that constitutive ERK1/2 activity in NSCLC cells promotes survival and chemotherapeutic survival [@PMID:12181740]. Additionally, we re-implemented these regression model with three different clustering strategies; (i) no clustering, (ii) k-means, and (iii) GMM using a complete portion of this data set (1.311 peptides). As expected, all strategies rendered excellent predictive performance due to the dramatic signaling differences between tumors and NATs (Supplemental Fig. 3A, C, and F). The fitting of a lasso model to the raw, unclustered, signaling data provides an alternative route to identify those particular phospho-peptides that contribute the most to explain the conserved signaling differences between tumors and NATs (e.g. SNTB2 S95-p, NICK1 Y105-p, MAP4 S636-p, IRF2BP2: S455-p;S457-p) (Supplemental Fig. 3B) and can be used as a strategy to complement the systematic approach that DDMC takes. Together, DDMC builds phosphoproteomic clusters that present signaling dysregulation common to tumors compared to NATs and identifies putative upstream kinases regulating them.  

![**Conserved tumor differences compared to normal adjascent tissue.** A) Hierarchical clustering of DDMC cluster centers. B–C) Principal components analysis scores (B) and loadings (C) of the samples and phosphopeptide clusters, respectively. D) Phospho-signal of tumor and NAT samples per cluster and statistical significance according to a Mann Whitney rank test. E) lasso receiver operating characteristic curve. F) lasso weights per cluster. G-I) Logo plot for the motifs of clusters 11 (G), 12 (H), and 23 (I). (J) Heatmap of Frobenius norms between the PSSMs of clusters 11, 12, and 23 and the PSPL matrices of putative upstream kinases.](figureM3.svg "Figure 3"){#fig:clusters width="100%"}

### Phosphoproteomic aberrations associated with genetic tumor differences

We then utilized the DDMC clusters to decipher the conserved signaling differences that differentiate LUAD patients harboring specific driver mutations. Reimplementing the same analysis framework as in Figure 3, we performed statistical testing and logistic regression to ascertain what particular clusters show the most prominent signaling changes between STK11 mutant and WT samples and enable the prediction of STK11's mutational status, respectively. The Mann Whitney tests indicated that most of the clusters were highly significant with most of their signal increased in mutant over WT samples. Among others, clusters 1, 7, 12, and 15 displayed an upregulation of phosphorylation signal in STK11 mutant samples versus WT whereas clusters 11 and 19 showed the opposite trend (Fig. 3A). We then utilized each patient's tumor and NAT phosphoproteomic data to succesfully predict the STK11 mutational status using a lasso model (Fig. 3B). The underlying signaling differences are mainly explained by the tumor samples of clusters 3, 8, 9, and 14, as well as by the NATs of clusters 2 and 17. These modeling results motivated further exploring clusters 1, 8, 9, and 15. The sequence motif of cluster 1 is enriched with basic residues at most positions, especially in the case of lysine at positions +2, +3, and +4. In addition, the presence of tyorsine at +1 in this motif could suggest that several peptides in this cluster may be coordinately primed by tyrosine kinases prior or after its central phosphorylation. There is also an overrepresentation of peptides displaying a cysteine in multiple positions, particularly at -3 and -2. Cluster 8 motif is characterized by a strong enrichment of hydrophobic residues in several positions such as tryptophan at +1 and +2, as well as leucine, isoluecine, phenylalanine, and methionine at -5, +4 and +5. In addition, this motif might also be jointly regulated by serine kinases at positions -1 and -2. Cluster 9 motif is almost exclusively enriched with hydrophobic amino acids and residues with polar uncharged side chains such as threonine and serine. Consequently, it shows a complete depletion of basic and acidic residues. The sequence motif of cluster 15 displays a general overrepresentation of acidic residues with a strong presence of glutamic acid at -1 and -2 and aspartic acid at +1. This motif is also defined by polar residues at the N-terminus and basic amino acids at its C-terminus (Fig. 4D-G). By comparing the PSSMs of these motifs to the specificity profiles of putative upstream kinases, we found that cluster 1 sequence closely resembles the optimal target motif of AMPK; cluster 8 might be modulated by several NEK kinases 4, 5, 6 and 8 as well as PKC-ζ and PKD; cluster 9 displays a motif favored by ERK2, and to a lesser extent NEK3 and DAPK3; and the motif of cluster 15 might be a good candidate for CK2 [**+ LITERATURE SEARCH AS IN FIGM3 / PSP SUBSTRATE ANALYSIS?**]. Clustering based on phosphorylation signal only via either k-means or GMM using the complete portion of the data set (1311 peptides—whereas DDMC is fitted to 30561 peptides) did not result in predictive models which indicates that incorporating the sequence information into the clustering criterion in this context improves the prediction of STK11 mutational status (Supplemental Figure 5A). 
We then investigated phosphoproteomic aberrations in patients harboring EGFR mutations and/or ALK fusions (EGFRm/ALKf). Except three clusters, all of them showed high statistical significance with a marked upregulation of phosphorylation signal in EGFRm/ALKf samples compared with their WT counterparts. Clusters with the highest signal increase in mutant samples include 1, 2, 8, and 12. Many of these were likely to show this contrasting signal due to the fact that most mutations where present in tumor samples wherein phosphorylation signal is substantially higher (Fig. 3). Moreover, a lasso model predicted clusters 1, 13, 20, and 23 to explain these signaling differences. Cluster 2 displays a motif with a noticeable enrichment for basic residues across several positions, especially histidine at position -2, and lysine at -1 and +1. Moreover, several peptides seem to show the hydrophobic residues methionine, leucine, valine, and phenylalanine at -4. In addition, there's a strong signal of the phosphorylatable residues tyrosine and threonine at position -3. The motif's C-terminus is characterized by a strong enrichment of phenylalanine at positions +2, +3. and +4. PKA, and to a lesser extent PKD and TβRII, are identified by the model as putative upstream kinase of cluster 2. Clusters 13 and 23 are clearly enriched with polar and hydrophobic residues, which as mentioned above, is a preferred motif for the pro-mitotic kinase ERK2. In addition, cluster 13 might ber modulated by NEK3 and DAPK3, and cluster 23 by DMPK1. Finally, cluster 20 shows an overall enrichment of acidic and hydrophobic residues, with a significant overrepresentation of isoleucine at several positions. AMPK, p38, and CK2 are identified as the corresponding upstream kinases (Supplemental Figure 4). In this case, the predictive performance of DDMC to predict EGFRm/ALKf mutational status was also superior to k-means but comparable to GMM (Supplemental Figure 5B). It is worth noting, however, that these methods were fitted to the complete portion of the data set that contains only 1311 whereas DDMC was fitted to 30561 peptides given its ability to handle missingness (Figure 2). Thus, DDMC provides insights into the mechanisms orchestrating tumorigenesis in patients for which a driver mutation is identified. 

![**Phosphoproteomic aberrations associated with STK11 mutational status.** A) Phospho-signal of STK11 WT and mutant samples per cluster and statistical significance according to a Mann Whitney rank test B) ROC of a lasso model predicting the STK11 mutational status and (C) its corresponding weights per sample type (D-G) Logo plots of PSSM motifs of clusters 1, 8, 9, and 15. H) Heatmap of putative upstream kinases.](figureM4.svg "Figure 4"){#fig:clusters width="100%"}

    
### Phosphoproteomic aberrations associated with clinical features

![**Figure 5.** Its caption.](figureM5.svg "Figure 5"){#fig:clusters width="100%"}

 #TODO in figures 3, 4, and 5:
- Benchmark upstream kinase predictions with:
    -> KSEA
    -> KinomeXplorer
    -> Scansite
