## Results

### Constructing an expectation-maximization algorithm tailored for clustering phosphoproteomic data

![**Schematic of the DDMC approach to cluster global signaling data and infer upstream kinases driving phenotypes.** A) DDMC simultaneously models phosphorylation signal and peptide sequence to cluster global signaling measurements. B) Predictive modeling using clusters allows one to establish assocations between specific clusters and features of interest. C) Putative upstream kinases regulating meaningful clusters can be predicted by computing the distance between a cluster motif's PSSM and a PSPL PSSM. PSSM; Position-specific scoring matrix, PSPL; Position scanning peptide library](DDMC_functional_diagram.svg "Figure 1"){#fig:method width="100%"}

MS-based global phosphoproteomic data provides unparalleled coverage when interrogating kinase signaling networks and their therapeutic implications. However, these data also present challenging issues as a consequence of their incomplete and stochastic coverage, high-content but low-sample throughput, and variation in coverage across experiments. In addressing these issues, we recognized that MS measurements simultaneously provide two pieces of information. Specifically, one obtains the exact site of phosphorylation on a peptide sequence and then some measure of abundance within the sample of interest. Both of these pieces of information are critical to the overall interpretation of the data.

Based on this observation, we built a mixture model that probabilistically clusters phosphosites based on both their peptide sequence and abundance across samples (Fig. {@fig:methodS}). In each iteration, DDMC applies an expectation-maximization algorithm to optimize clusters that capture the average features of member sequences and their abundance variation (Fig. {@fig:method}A and {@fig:methodS}). Clustering both the sequence and abundance measurements simultaneously ensures that the resulting clusters are a function of both types of features, which we hypothesized would provide both more meaningful and robust clusters.

The resulting clustering provides coordinated output that can be used in a few different ways. The cluster centers, by virtue of being a summary for the abundance changes of these peptides, can be regressed against phenotypic responses (e.g., cell phenotypes or clinical outcomes) to establish associations between particular clusters and response (Fig. {@fig:method}B). This can help to identify which cluster of peptides drives a certain response. Regression using the clusters instead of each peptide ensures that the model can be developed despite relatively few samples, with minimal loss of information since each peptide varies in a similar manner.
<!-- ASM: This data reduction also aids interpretation... -->

In parallel or independently, one can interrogate the resulting PSSMs to describe the overall sequence features of that cluster. These outputs can be readily compared to other information such as experimentally generated specificity profiles of putative upstream kinases via Position Specific Scanning Libraries (PSPL), to infer upstream kinases [@PMID:10945990; @PMID:20159853; @PMID:31124786; @PMID:15782149; @PMID:26551075]. We extracted a collection of 62 kinase specificity profiles to identify which cluster motifs most resemble the optimal motif of putative upstream kinases (Fig. {@fig:method}C) [@PMID:18765831; @PMID:15782149; @PMID:26551075]. However, kinase-substrate specificity is also dictated by features outside of the immediate substrate region, and we note that our approach is more general than strictly assembling kinase-substrate predictions. Overall, this overview demonstrates how DDMC can take complex, coordinated signaling measurements and find patterns in the phosphorylation signals to reconstruct signaling networks and associate particular clusters and phenotypes.

### A dual data-motif strategy improves prediction of different phenotypes

![**Sequence information enhances model prediction.** (A-O) Performance of a regression model predicting the mutational status of STK11 (A-E) EGFR and/or ALK (F-J) and tumor infiltration level (hot versus cold) (K-O) in LUAD patients using either only phosphorylation data (0), mainly peptide sequences (50), or a mix (15, 20, 40). EGFRm/ALKf; EGFR mutant and/or ALK fusion.](figureM2.svg "Figure 2"){#fig:preds width="100%"}

<!-- Intro to model set up and data -->
We utilized DDMC with the binomial sequence distance method and 24 clusters (Fig. {@fig:method}, {@fig:missingness}E/F). We were able to include many peptides that were not observed in every tumor through our ability to handle missing data, but still filtered out any peptides that were not captured in at least two 10-plex TMT runs, resulting in a data set of 30,561 peptides. We used this fitting result throughout the rest of this study. The resulting 24 cluster motifs can be found in Fig. {@fig:motifsS}.

<!-- Rationale and Results -->
To evaluate the benefit of incorporating the peptide sequence information into the clustering criterion, we asked whether utilizing DDMC with different sequence weights would affect the performance of a regularized logistic regression model that predicts the mutational status of STK11, whether a patients harbors a mutation in EGFR and/or a gene fusion in ALK (EGFRm/ALKf), and the level of tumor infiltration ("Hot" versus "Cold"). We found that for all three phenotypes, when the method does not consider the peptide sequences (weight=0), the patient samples are not as accurately classified as when a combination of both data and sequence is used (Fig. {@fig:preds}). Interestingly, in the case of STK11, the use of the largest weight wherein mainly the sequence motifs are used for clustering provided the best prediction performance (mean AUC=0.90±0.06) (Figs. {@fig:preds}A-E). Likewise, EGFRm/ALKf samples were best classified with a mix weight of 15 (mean AUC=0.84±0.07) or mainly prioritizing the sequence information (mean AUC=0.82±0.06) (Figs. {@fig:preds}F-J). Finally, the regression model classifying whether a sampkes is "hot-tumor-enriched" (HTE) or "cold-tumor-enriched" (CTE) showed its best fitness with a weight of 40 (AUC=0.82±0.04). Together, these results indicate that observing the motif information during clustering lead to final clusters that enhance the performance of regularized regression models predicting different phenotypes.

### Widespread, dramatic signaling differences between tumor and normal adjascent tissue

![**Conserved tumor differences compared to normal adjascent tissue.** A) Hierarchical clustering of DDMC cluster centers. B–C) Principal components analysis scores (B) and loadings (C) of the samples and phosphopeptide clusters, respectively. D) Phosphorylation signal of tumor and NAT samples per cluster and statistical significance according to a Mann Whitney rank test. E) lasso receiver operating characteristic curve. F) lasso weights per cluster. G-I) Logo plot for the motifs of clusters 11 (G), 12 (H), and 23 (I). (J) Heatmap of Frobenius norms between the PSSMs of clusters 11, 12, and 23 and the PSPL matrices of putative upstream kinases.](figureM3.svg "Figure 3"){#fig:TvsNAT width="100%"}

<!-- The signaling in both tissues is so different that any modeling strategy captures distinct patterns -->
We explored whether DDMC could recognize conserved signaling patterns in tumors compared to normal adjacent tissue (NAT). The signaling difference between tumors and NAT samples was substantial, highlighting the significant signaling rewiring that tumor cells must undergo (Fig. {@fig:TvsNAT}A). Using principal components analysis, we could observe that NAT samples were more similar to one another than each tumor sample (Fig. {@fig:TvsNAT}B/C). Nearly every cluster was significantly different in its average abundance between tumor and NAT (Fig. {@fig:TvsNAT}B/C). Not surprisingly given these enormous differences, samples could be almost perfectly classified using their phosphopeptide signatures, with or without DDMC (Fig. {@fig:TvsNAT}E; {@fig:TvsNATS}).

<!-- DDMC data -->
Hierarchical clustering of the DDMC cluster centers shows three distinct groups of clusters. One group had lower tumor abundances as compared to NATs (e.g., 11, 7, 19), another group showed the opposite (e.g., 1, 2, 12), and a few additional clusters either varied randomly (4, 21, 13) or showed inconsistent variation with tumor versus NAT status (9, 17, 20) (Fig. {@fig:TvsNAT}A). Consistent with the hierarchical clustering, the PCA loading of cluster 11 is negatively weighted along PC1, following all NAT samples (Fig. {@fig:TvsNAT}B). Conversely, clusters 1 and 12, among others, were positively weighted on PC 1 (Fig. {@fig:TvsNAT}C).

<!-- DDMC sequence -->
With the abundance changes we observed, we decided to further explore the peptide motifs of clusters 11, 12, and 23. Cluster 11 displays a strong conservation for lysine across all peptide positions, a clear enrichment of phenylalanine at -1, and an enrichment of tyrosine, lysine, methionine and histidine at +1 (Fig. {@fig:TvsNAT}G). In contrast, cluster 12 is enriched for acidic residues in most positions along with enrichment of tyrosine at -1, aspartic acid and tyrosine at +1, and lysine at +2 (Fig. {@fig:TvsNAT}H). The presence of tyrosines flanking the phosphorylation site of clusters 11 and 12 might suggest the activity of a coordinated regulation of this cluster wherein upon an initial phosphorylation event at either -1 or +1, the substrate may become activated for subsequent phosphorylation at the central serine [@PMID:26551075]. The sequence motif of cluster 23 is characterized by an overrepresentation of phenylalanine across several positions, especially at -3. By comparing the cluster PSSMs with our collection of specificity profiles of putative upstream kinases, we found that cluster 11 might be modulated by SLK and/or PAK2; cluster 12 by CK2, and cluster 23 by ERK2 (Fig. {@fig:TvsNAT}J). Taken together, DDMC builds phosphoproteomic clusters that present signaling dysregulation common to tumors compared to NATs and identifies putative upstream kinases modulating them. These features can help to interpret phosphoproteomic results, even when there are substantial differences between samples.

<!-- TODO: THE FACT THAT CLUSTERS 11 AND 23 ARE UPREGULATED IN NATs WOULD SUGGEST THAT SLK, PAK2, AND ERK2 ACT AS TUMOR SUPPRESSORS... Add to heatmap of upstream kinases all clusters with marked differential tumor vs NAT average abundance, not only 11, 12, and 23? General view may help explain what's going on -->

### Genetic driver mutations are associated with more targeted phosphoproteomic rewiring

![**Phosphoproteomic aberrations associated with STK11 mutational status.** A) Phosphorylation signal of STK11 WT and mutant samples per cluster and statistical significance according to a Mann Whitney rank test B) ROC of a lasso model predicting the STK11 mutational status and (C) its corresponding weights per sample type (D-G) Logo plots of PSSM motifs of clusters 1, 8, 9, and 15. H) Heatmap of putative upstream kinases.](figureM4.svg "Figure 4"){#fig:STK11 width="100%"}

<!-- STK11 DDMC data  -->
Inactivating somatic mutations in STK11 lead to increased tumorigenesis and metastasis [@PMID:17676035]. Thus, we aimed to identify the phosphoproteomic aberrations triggered by this genetic event. The majority of clusters were significantly altered, generally toward higher mutant abundances with a mutation (Fig. {@fig:STK11}A). The cluster centers corresponding to each patient's tumor and NAT samples could succesfully predict the STK11 mutational status by regularized logistic regression. The underlying signaling differences were mainly explained by the tumor samples of clusters 8 and 14 to which the model assigned very large negative weights (Fig. {@fig:STK11}B). These results motivated further exploration of clusters 8, 14, and 22 .

<!-- STK11 DDMC sequence -->
The motif of cluster 8 is characterized by a strong enrichment of hydrophobic residues in several positions such as tryptophan at +1 and +2, as well as leucine, isoluecine, phenylalanine, and methionine at -5, +4 and +5. In addition, this motif might also be jointly regulated by serine kinases at positions -1 and -2. Cluster 14 shows a PSSM motif with a clear enrichment of basic residues in its N-terminus, especially at positions -3 and -2. Also, hydrophobic residues seem to be favored at positions -4, -1 and +1, +4, and +5. The motif of cluster 22 displays a striking enrichment of all three basic residues at positions +1, +2, and +3, an overrepresentation of tyrosines at positions -5, -4, and to a lesser extent -2, and a strong presence of phenyalalanine at -2. (Fig. {@fig:STK11}D-F). By comparing the PSSMs of these motifs to the specificity profiles of putative upstream kinases, we found that cluster 1 sequence closely resembles the optimal target motif of AMPK; cluster 8 might be modulated by several NEK kinases 4, 5, 6 and 8 as well as PKC-ζ; cluster 14 displays a motif clearly favored by PKA and PKD, and to a lesser extent AKT; lastly, ERK1, NEK9, and SHC1 might be good candidates for the peptide motif of cluster 22 (Fig. {@fig:STK11}G). 

<!-- EGFRm/ALKf intro -->
Tyrosine kinase inihibitors (TKIs) targeting the receptor tyrosine kinases (RTKs) EGFR and ALK are effective treatments in cancer patients with EGFR mutations and/or ALK translocations. However, these treatments are limited by drug resistance which in some cases can be mediated by the concomitant signaling of both RTKs activated by driver mutations [@PMID:21791641; @PMID:27707887]. Therefore, we explored the conserved tumor signaling changes of patients harboring mutations in EGFR and/or gene fusions in ALK.

<!-- EGFRm/ALKf data -->
Except three clusters, all of them showed high statistical significance with a marked upregulation of phosphorylation signal in EGFRm/ALKf samples compared with their WT counterparts. Clusters with the highest signal increase in mutant samples include 1, 2, 8, and 12 (Fig. {@fig:EGFRm/ALKf}A). Moreover, a lasso model predicted clusters 1, 13, 20, and 23 to mostly explain these signaling differences (Fig. {@fig:EGFRm/ALKf}B/C). 

<!-- EGFRm/ALKf sequence -->
The sequence motif of cluster 1 is enriched with basic residues at most positions, especially in the case of lysine at positions +2, +3, and +4. In addition, the presence of tyorsine at +1 in this motif could suggest that several peptides in this cluster may be coordinately primed by tyrosine kinases prior or after its central phosphorylation. There is also an overrepresentation of peptides displaying a cysteine in multiple positions, particularly at -3 and -2. Clusters 13 and 23 are clearly enriched with polar and hydrophobic residues. As mentioned above, this feature is a preferred motif for the pro-mitotic kinase ERK2. Cluster 13 might be modulated by NEK3 and DAPK3. Finally, cluster 20 shows an overall enrichment of acidic and hydrophobic residues, with a significant overrepresentation of isoleucine at several positions which does not seem to provide an ideal docking environment for any of the candiate upstream kinases. Our model identifies AMPK and p38 as the upstream kinases for clusters 1 and 13, respectively (Fig. {@fig:EGFRm/ALKf}D-F and Fig {@fig:TvsNAT}I).

<!-- Comparisong to other regression strategies -->
Finally, we compared the classification performance of four regularized logistic regression models fit to either the DDMC clusters, clusters generated by the standard methods GMM and k-means, or the raw phosphoproteomic data directly. It is worth noting that unlike DDMC, GMM, k-means, and direct regression can not handle missing values and thus for this strategies we used the 1,311 peptides that were observed in all samples, whereas DDMC was fit to the entire data set comprising 30,561 phosphosites. Interestingly, we found that samples were classified with higher accuracy with DDMC than using GMM and similarly than with k-means, especially with STK11 (Fig {@fig:RegressionsS}A). Direct regression to the raw signaling data yielded excellent performance, however this strategy assigns thousands of coefficients to different peptides that vary every time the model is run which renders this appraoch unable to establish a consistent link between mutations and signaling (Fig {@fig:RegressionsS}).  


### Exploration of immune infiltration-associated signaling patterns in tumors   

![**Figure 5.** Phosphoproteomic signatures driving tumor immune infiltration.** (A-B) PCA scores (A) and loadings (B) using xCell on the RNA-seq data of tumors and NATs [@PMID:32649874; @PMID:29141660]. C) Multi-task lasso coefficients of the different immune population per cluster. (D-F) PSSM logo plots of cluster motifs 6 (D) and 9 (E), and its putative upstream kinases (F).](figureM5.svg "Figure 5"){#fig:TIIC width="100%"}

<!-- Intro to immunotherapy in NSCLC -->
Immune checkpoint inhibitors (ICIs) have emerged as effective treatment options for NSCLC patients. Unfortunately, there still is a need to inform patient selection and the most effective treatment program in different NSCLC clinical settings. Patients that do not respond to ICIs have tumors that present immune-cold behavior either inherently, or via an adaptive process after long exposure to the drug. However, the signaling mechanism by which malignant cells prevent the presence of tumor-infiltrating immune cells (TIICs) remains elusive. To this end, we used our DDMC clusters to explore the shared signaling patterns that differentiate HTE from CTE LUAD patient samples [@PMID:32649874; @PMID:29141660]. HTE and CTE status per patient was determined by Gilette et al upon use of xCell using RNA-seq of the tumor samples and subsequent consensus clustering [@PMID:32649874]. 

<!-- DDMC data -->
We observed four clusters that were statistically significantly different in its average abundance between HTE and CTE samples (Fig. {@fig:TvsNAT}A). Cluster 17, 18, and 20 display higher abundances in HTE with p-values lower than 0.05 and cluster 21 presents much lower average signal in HTE compared to CTE with a p-value lower than 0.01. Samples could be sucessfully classified using the signaling patterns captured by DDMC (Fig. {@fig:TvsNAT}B). The predictive performance of this regression was mainly explained by the positive and negative large weights assigned to cluster 6 and cluster 9, respectively. Additional clusters such as 10, 15, 16, 17, and 19 contributed to explain these signaling differences but to a lesser extent (Fig. {@fig:TvsNAT}C).

<!-- DDMC sequence -->
With the signaling changes observed in HTE compared to CTE samples and given the regression results described above, we further investigated the sequence motifs and putative upstream kinases of clusters 6, 9, 17, 18, and 21. Cluster 6 presents a strong enrichments of hydrophobic residues—especially tryptohan—in its N-terminus as well as at both positions adjacent to the target site. In addition, tyrosine is enriched at position +3, and cysteine at positions +4 and +5. Interestingly, there seems to be a contrasting enrichment of basic residues in its N-terminus versus a preference for acidic residues in C-terminus (Fig. {@fig:motifsS}F). Cluster 9 is exclusively represented by the polar residues serine, threonine, and cysteine, and the hydrophobic residues proline, alanine, and methionine (Fig. {@fig:motifsS}I). Cluster 17 shows an enrichment for the hydrophobic residues valine, tryptophan, or phenylalanine at positions -5, -4, -2, and +4, glutamic acid and tyrosine at -1, and the polar residues cysteine and glycine at +1. Similarly to cluster 9, clusters 18 and 21 also show an exclusive overrepresentation of polar and hydrophobic amino acids. However, in these clusters the enriched residues in each biochemical group are more varied (Fig. {@fig:TvsNAT}D-F). By comparing the PSSMs of these cluster motifs to the specificity profiles of upstream kinases, we found that cluster 6 might be modulated by PKA and PKD; cluster 9 by ERK2; cluster 17 by AMPK; cluster 18 by ERK2 and DAMPK1; cluster 20 by AMPK; and cluster 21 also by ERK2. All in all, this analysis demonstrates that DDMC can help uncover common signaling patterns that explain differences between LUAD patients presenting HTE and CTE tumors. 

<!-- TODO: STK11 signaling inducing an immune-cold behavior  -->
