## Results

### A dual data-motif clustering strategy improves phosphoproteomic clustering

![Its caption.](figureM1.svg "Figure 1"){#fig:method width="100%"}
- Model diagram (maybe use entire figure) –– DONE

Figure 2 
- Diagram of aritifical missingness 
- Artifical missingness error and wins for different missingness percentages –– DONE
- Error across #clusters (+baseline, imputed average, and also maybe random clusters?)
- Error accross different weights (-baseline, +baseline supplemental)
- Regression performance comapring imputed average vs GMM alone vs co-clustering. Take patients with ALK fusions, peptides related have missingess.

### Common tumor-NAT differences

![**Figure 2.** Its caption.](figureM2.svg "Figure 2"){#fig:tn width="100%"}
- Clustermap of clusters
- PCA analysis of clusters
- Regression to predict tumor vs NAT––Logistic Regression, pick out most significant coefficients. 
- Enrichment / Sequence analysis of interesting clusters
    -> KSEA
    -> Motif-logos to explore upstream kinases

### Phosphoproteomic aberattions associated with molecular and histological signatures

![**Figure 3.** Its caption.](figureM3.svg "Figure 3"){#fig:clusters width="100%"}
- **Definine patient clusters. What's unique about those clusters.** 
- Potentially interesting molecular features:
    -> Alk fusions
    -> Copy number alteration
    -> DNA methylation––CIMP
    -> Somatic mutations
    -> Immune landscape ––> xCell Aran et all 2017

    
### Phosphoproteomic aberattions associated with clinical features
- Regression against survival (additional clinical features...?, other LUAD large-scale signaling set?)
- Potentially interesting histological features:
    -> LUAD histological subtype, focality, tumor site, stage...
- Enrichment / Sequence analysis of interesting clusters
    -> KSEA
    -> Motif-logos to explore upstream kinases