## Results


### Constructing an expectation-maximization algorithm tailored to cluster phosphoproteomic data

MS-based global phosphoproteomic data provides unparalleled coverage to interrogate signaling networks and its therapeutic implications. However, these data present challenging issues such as undersampling of the phosphoproteome and its high-content but low-sample-throughput nature. To tackle these issues, we have built an expectation-maximization (EM) algorithm that probabilistically and simultaneously models the peptide phosphorylation levels with a Gaussian Mixture Model (GMM) and the peptide sequences using either a PAM250 transition matrix, or a binomial position enrichment model (Fig. 1 and methods section). The user-defined variables of this algorithm are the number of clusters the scaling factor or weight to prioritize the data or the sequence scores. During the Expectation– or E-step, the algorithm guesses a *data* and a *sequence* probability distribution given the current model parameters to assign every peptide into a cluster based on its maximum likelihood estimation. That is, the GMM generates the probabilities of each phosphosite being assigned to every cluster according to its log-transformed abundance. In the PAM250 method, the average distance between a peptide's sequence and each of those within a cluster are calculated using the transition matrix. For the binomial method, we collected 333,492 random MS-generated phosphopeptides from the public compendium PhosphoSitePlus [@PMID:25514926] to be used as the set of background sequences. The probability of observing each amino acid at each position was calculated by comparing each cluster and the background using an incomplete beta function. First, the cluster sequences are used to build counts matrices displaying the number of amino acid occurrences across positions. These matrices are then individually compared to the weighted frequencies matrix of the background set of sequences using a binomially distributed model in order to generate a binomial probability matrix for each cluster. The probability of a peptide being assigned to any cluster is the average of each position's probability present in the correspondinog cluster's binomial probability matrix. For each cluster, the final score is the sum of the GMM and sequence scores multiplied by a user-defined weight which scales both scores so that a weight of 0 or a very large number only scores the data or sequence, respectively. Importantly, the final scores represent the relative probabilities that define the partial membership of the query peptide to each cluster. In other words, the EM algorithm provides soft cluster assignments for every peptide. Then, the GMM parameters and cluster sequences are updated during the M-step based on the assignments generated during the previous E-step. These two steps repeat until convergence.


### Dual data-motif clustering strategy improves phosphoproteomic clustering

![Its caption.](figureM1.svg "Figure 1"){#fig:method width="100%"}
- Model diagram (maybe use entire figure) –– DONE

Figure 2 
- Diagram of aritifical missingness 
- Artifical missingness error and wins for different missingness percentages –– DONE
- Error across #clusters (+baseline, imputed average, and also maybe random clusters?)
- Error accross different weights (-baseline, +baseline supplemental)
- Regression performance comapring imputed average vs GMM alone vs co-clustering. Take patients with ALK fusions, peptides related have missingess.

### Common tumor-NAT differences

![**Figure 2.** Its caption.](figureM2.svg "Figure 2"){#fig:tn width="100%"}
- Clustermap of clusters
- PCA analysis of clusters
- Regression to predict tumor vs NAT––Logistic Regression, pick out most significant coefficients. 
- Enrichment / Sequence analysis of interesting clusters
    -> KSEA
    -> Motif-logos to explore upstream kinases

### Phosphoproteomic aberattions associated with molecular and histological signatures

![**Figure 3.** Its caption.](figureM3.svg "Figure 3"){#fig:clusters width="100%"}
- **Definine patient clusters. What's unique about those clusters.** 
- Potentially interesting molecular features:
    -> Alk fusions
    -> Copy number alteration
    -> DNA methylation––CIMP
    -> Somatic mutations
    -> Immune landscape ––> xCell Aran et all 2017

    
### Phosphoproteomic aberattions associated with clinical features
- Regression against survival (additional clinical features...?, other LUAD large-scale signaling set?)
- Potentially interesting histological features:
    -> LUAD histological subtype, focality, tumor site, stage...
- Enrichment / Sequence analysis of interesting clusters
    -> KSEA
    -> Motif-logos to explore upstream kinases