## Results


### Constructing an expectation-maximization algorithm tailored to cluster phosphoproteomic data

MS-based global phosphoproteomic data provides unparalleled coverage to interrogate signaling networks and its therapeutic implications. However, these data present challenging issues such as undersampling of the phosphoproteome, its high-content but low-sample-throughput nature, and the presence of missing values in large-scale phosphoproteomics data sets. To tackle these issues, we have built an expectation-maximization (EM) algorithm that probabilistically and simultaneously models the peptide phosphorylation levels with a Gaussian Mixture Model (GMM) and the peptide sequences using either a PAM250 transition matrix, or a binomial position enrichment model (Supplemental Fig. 1). In each iteration, DDMC's recursive algorithm optimizes the clusters of phosphopeptides until the peptides included in each final cluster displays conserved kinase motifs in addition to similar phosphorylation behaviors (Fig. 1A and Supplemental Fig. 1). We hypothesized that this dual clustering criteria will provide the user with more biologically meaningul clusters than with either sequence or data alone since, given the relevance of kinase specificity in signaling networks, the peptide sequences provide additional useful information into the global signaling context. Next, the resulting cluster centers can be used to fit regression models and predict features of interest (e.g. phenotyes or clinical outcomes) in order to establish associations between particular clusters and such features (Fig. 1B). This helps identify what specific clusters are meaningful in a specific study versus those that are not. For instance, as illustreated by Fig. 1C, clusters 1, 2, and 3 each explain a particular phenotye along PC1 and PC2, whereas cluster 4 may not be as directly implicated as their counterparts. Finally, the incorporation of the sequence information into the model and the subsequent generation of representative cluster motifs provides the opportunity of quantifying the similarity between a cluster's position-specific-scoring matrix (PSSM) and the specificity profiles of putative upstream kinaes, also in the form of a PSSM, generated via Position Specific Scanning Libraries (PSPL). In this biochemical assay, a kinase of interest is incubated with disparate biotinylated peptide libraries to find the "optimal" substrate motif [CITE]. We extracted a collection of 62 kinase specificity profiles to find what cluster motifs resemble most accuratly the optimal motif of putative upstream kinases. To do so, we compute the Frobenius norm for all pairwise combinations between cluster PSSMs and PSPL PSSMs to predict what upstream kinases might be modulating specific clusters (Fig 1C). Hence, DDMC takes complex, coordinated signaling measurements and finds patterns in phosphorylation signal as well as sequence similarity to reconstruct signaling networks and enable the association between particular clusters and cell phenotypes. 


![**Schematic of the DDMC approach to cluster global signaling data and infer upstream kinases driving phenotypes.** A) DDMC simultaneously models phosphorylation signal and peptide sequence to cluster global signaling measurements. B) Modeling based on clusters such as regression methods allows establisheing assocations between specific clusters and features of interest. C) Putative upstream kinases regulating meaningful clusters can be predicted by computing the distance between a cluster motif's PSSM and a PSPL PSSM. PSSM; Position-specific scoring matrix, PSPL; Position scanning peptide library](EM_diagram.svg "Figure 1"){#fig:method width="100%"}


### Dual data-motif clustering strategy robustly imputes missing values

To evaluate the robustness of our combinad dual data-motif clustering (DDMC) method in analyzing incomplete data sets, we designed a computational experiment wherein we removed specific observations and attempted to predict them using the cluster centers corresponding to the peptides those missing values belonged to (Figure 2A). The resulting mean squared errors between the actual and predicted values were compared to compared to commonly used imputation strategies such as the peptides' mean or minimum signal. Furthermore, we evaluated the imputation performance of our method when clustering the data by phosphorylation signal only, by sequence similarity only, or by combining both information sources (2B-D). These results indicate that MDCC clearly imputes missing values more accurately than the peptides' mean signal in all three cases. No noticeable differences were observed in the imputations performance using data or sequence only, or a mix of both. Next, we re-implemented this strategy to find out what number of clusters (2E) and weight (2F) yielded more accurate predictions. We found that a larger number of clusters and smaller weights (prioritizing phospho-signal over peptide sequence) minimized the imputation error. These data suggests that DDMC outperforms standard imputation strategies such as peptide mean signal (and minimum signal).   

#TODO: 
- Compare to minimum signal
- Filling the missingness with 0's
- Matrix completion by PCA
- Add legend for each regression line in plots B-F and mix labels/titles

![**Benchmarking the robustness of motif clustering to missing measurements.** A) A schematic of the process for quantifying robustness to missing values. Any peptides containing less than 7 TMT experiments were discarded. For the remaining 15904 peptides, an entire random TMT experiment was removed per peptide and these values were stored for later comparison. Next, these artifical missing values were imputed using either a baseline strategy (peptide mean/minimum signal) or the corresponding cluster center. Once a mean squared error was computed for each peptide, the second iteration repeats this process by removing a second TMT experiment. A total of 5 random TMT experiments per peptide were imputed by clustering using either the peptides' phosphorylation signal only (B), the peptides' sequence only (C) or a mix of both information sources (D). The imputation errors were calculated across different number of clusters (E) and different data-vs-sequence weights (F).](figureM2.svg "Figure 2"){#fig:tn width="100%"}


### Shared differences between tumor and normal adjascent tissue

We explored whether DDMC could recognize conserved signaling patterns in tumors compared with normal adjacen tissue (NAT) samples. To do so, we fit DDMC using the binomial sequence distance method with 24 clusters and a weight of 15 (Fig. 1, 2E, F). We used this fitting result for every analysis performed in this study. Hierarchical clustering of the cluster centers shows a noticeable contrasting signal between tumor and NAT samples (Fig 3A). This heatmap shows two distinctly regulated groups of NAT samples (top and bottom) whereas all tumor samples cluster together. Three signaling distinct patterns can be observed; (i) high signal in tumors and low in both NAT groups, (ii) heterogeneous signal in tumors, one high NAT group (top) and one low NAT group (bottom), and (iii) cluster 11 shows a unique signaling wherein these peptides are enriched in most NAT samples and depleted in all tumors. The scores of the PCA analysis of the cluster centers further underscore the striking signaling differences between tumor—positively weighted across principal component (PC) 1—and NAT samples—negatively weighted across PC1—(Fig. 3B). Consistent with the hierarchical clustering, the PCA loading of cluster 11 is negatively weighted along PC1, closely tracking with all NAT samples. Conversely, clusters 12 and 1 display similar PC1 weights as tumor samples (Fig. 3C). Additionally, we asked whether for every cluster the signaling between tumor versus NAT samples were statistically different. We performed Mann Whitney rank tests and found that all clusters display significantly different phosphorylation levels in their tumor samples compared with NATs (Fig. 3C&D). As expected, the overall signal of tumors is strikingly lower than in NAT samples in cluster 11, whereas cluster 12 shows the opposite pattern. Next, we fit a lasso logistic regression model to predict sample type (tumor or NAT) using the cluster centers. All samples were correctly predicted except from one, again highlighting the striking differences in signaling between cancer and normal cells (Fig. 3E&F). Most of these signaling differences were expleined by cluster 11 (Fig. G). With this information, we decided to further explore clusters 11, 12, and 23. By looking at what residues where enriched across all peptide positions, we found that cluster 11 contains phospho-serine peptides only with a strong conservation for K in the motif's C-terminus, a clear enrichment of F at -1, and a mix enrichment of Y, K, M and H at +1 (3H). Overall, these cluster shows a clear enrichment of basic residues across most of the peptide positions. In contrast, cluster 12—also containing phospho-serine peptides only—displays a motif with a clear enrichment for acidic residues. In addition, this cluster shows a large enrichment of Y at -1 and K at +2 (3I). By comparing the cluster PSSMs with our collection of specificity profiles of putative upstream kinases, we found that cluster 11 might be modulated by SLK and PAK2; cluster 12 by CKII, SRC, and TβRII, and cluster 23 by ERK2, and DMPK1 (Fig. 3J). Additionally, we re-implemented these regression model with three different clustering strategies; (i) no clustering, (ii) k-means, and (iii) GMM. As expected due to the dramatic signaling differences between tumors and NATs, all strategies rendered excellent predictive performance (Supplemental Fig. 3A,C&F). Interestingly, the lasso model fitted to the unclustered data identified 10 p-sites that mostly explain the conserved tumor signaling differences compared to NATs. GPBP1L1 S76-p, MAP4 S636-p, SSH3 S37-p, NCK1 Y105-p, and SNTB2 S95-p expalin these differences in tumors whereas ZYX S344-p, SPECC1L S220-p, IRF2BP2 S455-p S457-p, CCDC102A S219-p, and THRAP3 S253-p in NAT samples (Supplemental Figure 3)[**NOTE THAT THE LASSO MODEL IS IRREPRODUCIBLE WITH UNCLUSTERED DATA: DIFFERENT WEIGTHED P-SITES IN EVERY FITTING**]. 

![**Conserved tumor differences compared to normal adjascent tissue.** A) Hierarchical clustering of DDMC cluster centers. B–C) Principal components analysis scores (B) and loadings (C) of the samples and phosphopeptide clusters, respectively. D) Phospho-signal of tumor and NAT samples per cluster and statistical significance according to a Mann Whitney rank test. E) lasso receiver operating characteristic curve. F) lasso weights per cluster. G-I) Logo plot for the motifs of clusters 11 (G), 12 (H), and 23 (I). (J) Heatmap of Frobenius norms between the PSSMs of clusters 11, 12, and 23 and the PSPL matrices of putative upstream kinases.](figureM3.svg "Figure 3"){#fig:clusters width="100%"}

### Phosphoproteomic aberrations associated with genetic tumor differences

We then used the DDMC clusters to decipher the conserved signaling differences according to the genetic background among LUAD patients. Adapting the same analysis framework as in Figure 3, we performed statistical testing and logistic regression to ascertain what particular clusters show the most prominent signaling changes between STK11 mutant and WT samples and enable the prediction of STK11's mutational status, respectively. The Mann Whitney tests indicated that most of the clusters were highly significant with most of their signal increased in mutant over WT samples. Clusters 1, 7, 12, and 15 displayed a striking upregulation of phosphorylation signal in STK11 mutant samples versus WT whereas clusters 11 and 19 showed the opposite trend (Fig. 3A). We then utilized each patient's tumor and NAT phosphoproteomic data to succesfully predict the STK11 mutational status through a lasso model (Fig. 3B). The underlying signaling differences are mainly explained by the tumor samples of clusters 8, 9, and 14, as well as by both tumor and NAT samples of cluster 22. The tumor samples of clusters 8 and 14 show a substantial downregulation in phosphorylation signal whereas peptides in cluster 9 are enriched in tumor samples. Lastly, cluster 22 is positively weighted in tumor samples and negatively weighted in NATs. These modeling results motivated further exploring clusters 1, 8, 9, and 15. The sequence motif of cluster 1 is enriched with basic residues throughout most of the peptide's positions, especially in position +2, +3, and +4. In addition, the presence of tyorsine at +1 in this motif could suggest that his target site might be previously or subsequently phosphorylated at this position by tyrosine kinases. There is also an overrepresentation of peptides displaying a cysteine in multiple positions, particularly in positions -3 and -2. Cluster 8 motif is characterized by a strong enrichment of hydrophobic residues in several positions such as tryptophan at +1 and +2, as well as leucine, isoluecine, and phenylalanine at +4 and +5. In addition, this motif might also be regulated by serine kinases at positions -1 and -2. Cluster 9 motif is almost exclusively enriched with hydrophobic amino acids and residues with polar uncharged side chains such as threonine and serine. Consequently, it shows a complete depletion of basic and acidic residues. The sequence motif of cluster 15 displays a general overrepresentation of acidic residues with a strong presence of glutamic acid at -1 and -2 and aspartic acid at +1. This motif is also defined by polar residues at the N-terminus and basic amino acids at its C-terminus (Fig. 4D-G). All of these clusters have a serine at the phosphoacceptor position 0. By comparing the PSSMs of these motifs to the specificity profiles of putative upstream kinases, we found that cluster 1's sequence most closely resembles the optimal target motif of AMPK, CDK5, and SHC1; cluster 8 might be modulated by several NEK kinases 4, 5, 6 and 8 as well as PKC-ζ and PKD; cluster 9 displays a motif strikingly favored by ERK2, and to a lesser extent NEK3 and DAPK3; and the motif of cluster 15 might be a good candidate to ve targeted by CKII and the receptor TβRII. Clustering based on phosphorylation signal only via either k-means or GMM using the complete portion of the data set (1311 peptides—whereas DDMC is fitted to 30561 peptides) did not result in predictive models which indicates that incorporating the sequence information into the clustering criterion in this context improves the prediction of STK11 mutational status (Supplemental Figure 5A). 
We reimplemented this modeling strategy to decipher the phosphoproteomic aberrations in patients harboring EGFR mutations and/or ALK fusions (EGFRm/ALKf). Except three clusters, all of them showed high statistical significance with a marked upregulation of phosphorylation signal in EGFRm/ALKf samples compared with their WT counterparts. Clusters with the highest signal increase in mutant samples include 1, 2, 8, and 12. Many of these were likely to show this contrasting signal due to the fact that most mutations where present in tumor samples wherein phosphorylation signal is substantially higher (Fig. 3). Moreover, a lasso model predicted clusters 1, 13, 20, and 23 to explain these signaling differences. Cluster 2 displays a motif with a noticeable enrichment for basic residues across several positions, especially histidine at position -2, and lysine at -1 and +1. Moreover, several peptides seem to show the hydrophobic residues methionine, leucine, valine, and phenylalanine at -4. In addition, there's a strong signal of the phosphorylatable residues tyrosine and threonine at position -3. The motif's C-terminus is characterized by a strong enrichment of phenylalanine at positions +2, +3. and +4. PKA, and to a lesser extent PKD and TβRII, are identified by the model as putative upstream kinase of cluster 2. Clusters 13 and 23 are clearly enriched with polar and hydrophobic residues, which as mentioned above, is a preferred motif for the pro-mitotic kinase ERK2. In addition, cluster 13 might ber modulated by NEK3 and DAPK3, and cluster 23 by DMPK1. Finally, cluster 20 shows an overall enrichment of acidic and hydrophobic residues, with a significant overrepresentation of isoleucine at several positions. AMPK, p38, and CKII are identified as the corresponding upstream kinases (Supplemental Figure 4). In this case, the predictive performance of DDMC to predict EGFRm/ALKf mutational status was also superior to k-means but comparable to GMM (Supplemental Figure 5B). It is worth noting, however, that these methods were fitted to the complete portion of the data set that contains only 1311 whereas DDMC was fitted to 30561 peptides given its ability to handle missingness (Figure 2). 

![**Phosphoproteomic aberrations associated with STK11 mutational status.** A) Phospho-signal of STK11 WT and mutant samples per cluster and statistical significance according to a Mann Whitney rank test B) ROC of a lasso model predicting the STK11 mutational status and (C) its corresponding weights per sample type (D-G) Logo plots of PSSM motifs of clusters 1, 8, 9, and 15. H) Heatmap of putative upstream kinases.](figureM4.svg "Figure 4"){#fig:clusters width="100%"}

    
### Phosphoproteomic aberrations associated with clinical features

![**Figure 5.** Its caption.](figureM5.svg "Figure 5"){#fig:clusters width="100%"}

 #TODO in figures 3, 4, and 5:
- Benchmark upstream kinase predictions with:
    -> KSEA
    -> KinomeXplorer
    -> Scansite
