## Results


### Constructing an expectation-maximization algorithm tailored to cluster phosphoproteomic data

MS-based global phosphoproteomic data provides unparalleled coverage to interrogate signaling networks and its therapeutic implications. However, these data present challenging issues such as undersampling of the phosphoproteome, its high-content but low-sample-throughput nature, and the presence of missing values in large-scale phosphoproteomics data sets. To tackle these issues, we have built an expectation-maximization (EM) algorithm that probabilistically and simultaneously models the peptide phosphorylation levels with a Gaussian Mixture Model (GMM) and the peptide sequences using either a PAM250 transition matrix, or a binomial position enrichment model (Fig. 1 and methods section). The user-defined variables of this algorithm are the number of clusters the scaling factor or weight to prioritize the data or the sequence scores. During the Expectation– or E-step, the algorithm guesses a *data* and a *sequence* probability distribution given the current model parameters to assign every peptide into a cluster based on its maximum likelihood estimation. That is, the GMM generates the probabilities of each phosphosite being assigned to every cluster according to its log-transformed abundance. In the PAM250 method, the average distance between a peptide's sequence and each of those within a cluster are calculated using the transition matrix. For the binomial method, we collected 333,492 random MS-generated phosphopeptides from the public compendium PhosphoSitePlus [@PMID:25514926] to be used as the set of background sequences. The probability of observing each amino acid at each position was calculated by comparing each cluster and the background using an incomplete beta function. First, the cluster sequences are used to build counts matrices displaying the number of amino acid occurrences across positions. These matrices are then individually compared to the weighted frequencies matrix of the background set of sequences using a binomially distributed model in order to generate a binomial probability matrix for each cluster. The probability of a peptide being assigned to any cluster is the average of each position's probability present in the correspondinog cluster's binomial probability matrix. For each cluster, the final score is the sum of the GMM and sequence scores multiplied by a user-defined weight which scales both scores so that a weight of 0 or a very large number only scores the data or sequence, respectively. Importantly, the final scores represent the relative probabilities that define the partial membership of the query peptide to each cluster. In other words, the EM algorithm provides soft cluster assignments for every peptide. Then, the GMM parameters and cluster sequences are updated during the M-step based on the assignments generated during the previous E-step. These two steps repeat until convergence.

![**Schematic of the DDMC simultaneous data and peptide sequence clustering approach.** Peptides are initialized into clusters at random. The measurements of abundance are represented by a multivariate Gaussian mixture model where each dimension of the distribution represents the abundance within a sample. Next, an expectation-maximization fitting scheme is used. During the expectation step, the distance of each peptide sequence to each cluster is calculated. This is done either through a binomial enrichment scheme (method 1) or using the average PAM250 distance (method 2). In parallel, the distance of each peptide abundance is compared to the cluster centers. These two distances are combined to update the assignments of each peptide to each cluster. During the maximization step, the cluster centers of the data are updated based on the weighted average of the peptide abundances in each condition. The peptide motifs are similarly updated through a weighted combination of the assigned peptides. Both steps continue sequentially until the change in peptide assignments between each iteration drops below a threshold.](EM_diagram.svg "Figure 1"){#fig:method width="100%"}


### Dual data-motif clustering strategy improves phosphoproteomic clustering

To evaluate the robustness of our combinad dual data-motif clustering (DDMC) method in analyzing incomplete data sets, we designed a computational experiment wherein we removed specific observations and attempted to predict them using the cluster centers corresponding to the peptides those missing values belonged to (Figure 2A). The resulting mean squared errors between the actual and predicted values were compared to compared to commonly used imputation strategies such as the peptides' mean or minimum signal. Furthermore, we evaluated the imputation performance of our method when clustering the data by phosphorylation signal only, by sequence similarity only, or by combining both information sources (2B-D). These results indicate that MDCC clearly imputes missing values more accurately than the peptides' mean signal in all three cases. No noticeable differences were observed in the imputations performance using data or sequence only, or a mix of both. Next, we re-implemented this strategy to find out what number of clusters (2E) and weight (2F) yielded more accurate predictions. We found that a larger number of clusters and smaller weights (prioritizing p-signal over peptide sequence) minimized the imputation error. These data suggests that DDMC outperforms standard imputation strategies such as peptide mean signal (and minimum signal).   

#TODO: 
- Compare to minimum signal
- Compare to GMSimpute/DreamAI
- Add legend for each regression line in plots B-F and mix labels/titles

![**Benchmarking the robustness of motif clustering to missing measurements.** A) A schematic of the process for quantifying robustness to missing values. Any peptides containing less than 7 TMT experiments were discarded. For the remaining 15904 peptides, an entire random TMT experiment was removed per peptide and these values were stored for later comparison. Next, these artifical missing values were imputed using either a baseline strategy (peptide mean/minimum signal) or the corresponding cluster center. Once a mean squared error was computed for each peptide, the second iteration repeats this process by removing a second TMT experiment. A total of 5 random TMT experiments per peptide were imputed by clustering using either the peptides' phosphorylation signal only (B), the peptides' sequence only (C) or a mix of both information sources (D). The imputation errors were calculated across different number of clusters (E) and different data-vs-sequence weights (F).](figureM2.svg "Figure 2"){#fig:tn width="100%"}


### Shared differences between tumor and normal adjascent tissue

![**Conserved tumor differences compared to normal adjascent tissue.** A) AAA. B–C) Principal components analysis scores (B) and loadings (C) of the samples and phosphopeptide clusters, respectively.](figureM3.svg "Figure 3"){#fig:clusters width="100%"}
- Clustermap of clusters
- PCA analysis of clusters
- Regression to predict tumor vs NAT––Logistic Regression, pick out most significant coefficients. 
- Enrichment / Sequence analysis of interesting clusters
    -> KSEA (Supplemental)
    -> Motif-logos to explore upstream kinases

### Phosphoproteomic aberations associated with genetic tumor differences

![**Phosphoproteomic aberations associated with genetic tumor differences.** Its caption.](figureM4.svg "Figure 4"){#fig:clusters width="100%"}

- **Definine patient clusters. What's unique about those clusters.** 
- Potentially interesting molecular features:
    -> Alk fusions
    -> Copy number alteration
    -> DNA methylation––CIMP
    -> Somatic mutations
    -> Immune landscape ––> xCell Aran et all 2017

    
### Phosphoproteomic aberations associated with clinical features

![**Figure 5.** Its caption.](figureM5.svg "Figure 5"){#fig:clusters width="100%"}
- Regression against survival (additional clinical features...?, other LUAD large-scale signaling set?)
- Potentially interesting histological features:
    -> LUAD histological subtype, focality, tumor site, stage...
- Enrichment / Sequence analysis of interesting clusters
    -> KSEA
    -> Motif-logos to explore upstream kinases
