## Results

### Constructing an expectation-maximization algorithm tailored for clustering phosphoproteomic data

![**Schematic of the DDMC approach to cluster global signaling data and infer upstream kinases driving phenotypes.** A) DDMC simultaneously models phosphorylation signal and peptide sequence to cluster global signaling measurements. B) Predictive modeling using clusters allows one to establish assocations between specific clusters and features of interest. C) Putative upstream kinases regulating meaningful clusters can be predicted by computing the distance between a cluster motif and PSPL PSSM. PSSM; Position-specific scoring matrix, PSPL; Position scanning peptide library](DDMC_functional_diagram.svg "Figure 1"){#fig:method width="100%"}

MS-based global phosphoproteomic data provides unparalleled coverage when interrogating kinase signaling networks and their therapeutic implications. However, these data also present challenging issues as a consequence of their incomplete and stochastic coverage, high-content but low-sample throughput, and variation in coverage across experiments. In addressing these issues, we recognized that MS measurements simultaneously provide two pieces of information. Specifically, one obtains the exact site of phosphorylation on a peptide sequence and then some measure of abundance within the measured samples. Both of these pieces of information are critical to the overall interpretation of the data.

Based on this observation, we built a mixture model that probabilistically clusters phosphosites based on both their peptide sequence and abundance across samples (Figure {@fig:methodS}). In each iteration, DDMC applies an expectation-maximization algorithm to optimize clusters that capture the average features of member sequences and their abundance variation (Figure {@fig:method}A and {@fig:methodS}). Both information sources—data and sequence—can be distinclty prioritized by defining the 'weight' parameter. With a weight of 0, DDMC is essentially a Gaussian Mixture Model (GMM) that clusters peptides based on the phosphorylation signal only whereas with a very large weight DDMC only uses the peptide sequences as clustering criterion. Clustering both the sequence and abundance measurements ensures that the resulting clusters are a function of both features, which we hypothesized would provide both more meaningful and robust clusters.

The resulting clustering provides coordinated output that can be used in a few different ways. The cluster centers, by virtue of being a summary for the abundance changes of these peptides, can be regressed against phenotypic responses (e.g., cell phenotypes or clinical outcomes) to establish associations between particular clusters and response (Figure {@fig:method}B). This can help to identify which cluster of peptides drives a certain response. Regression using the clusters instead of each peptide ensures that the model can be developed despite relatively few samples, with minimal loss of information since each peptide within a cluster varies in a similar manner.

In parallel or independently, one can interrogate the resulting position-Specific Scoring Matrices (PSSMs) to describe the overall sequence features of that cluster. These outputs can be readily compared to other information such as experimentally generated specificity profiles of putative upstream kinases via Position Specific Scanning Libraries (PSPL), to infer upstream kinases [@PMID:10945990; @PMID:20159853; @PMID:31124786; @PMID:15782149; @PMID:26551075]. We extracted a collection of 62 kinase specificity profiles to identify which cluster motifs most resemble the optimal motif of putative upstream kinases (Figure {@fig:method}C) [@PMID:18765831; @PMID:15782149; @PMID:26551075]. However, kinase-substrate specificity is also dictated by features outside of the immediate substrate region, and we note that our approach is more general than strictly assembling kinase-substrate predictions. Overall, this overview demonstrates how DDMC can take complex, coordinated signaling measurements and find patterns in the phosphorylation signals to reconstruct signaling networks and associate particular clusters and phenotypes.


 ### Dual data-motif clustering strategy robustly imputes missing values

![**Benchmarking the robustness of motif clustering to missing measurements.** A) A schematic of the process for quantifying robustness to missing values. Any peptides containing less than 7 TMT experiments were discarded. For the remaining 15904 peptides, an entire random TMT experiment was removed per peptide and these values were stored for later comparison. Next, these artifical missing values were imputed using either a baseline strategy (peptide mean/minimum signal, zeros, matrix completion by PCA) or the corresponding cluster center. Once a mean squared error was computed for each peptide, the second iteration repeats this process by removing a second TMT experiment. A total of 5 random TMT experiments per peptide were imputed by clustering using a different number of clusters (B-E) or different weights (E-I).](figureM2.svg "Figure 2"){#fig:missingness width="100%"}

To evaluate the robustness of our combinad dual data-motif clustering (DDMC) method in analyzing incomplete data sets, we designed a computational experiment wherein we removed specific observations and attempted to predict them using the cluster centers corresponding to the peptides those missing values belonged to (Figure {@fig:missingness}A). The resulting mean squared errors between the actual and predicted values were compared to commonly used imputation strategies such as the peptides' mean or minimum signal, zeros, or matrix completion by PCA. Furthermore, we evaluated the imputation performance of our method when clustering the data using a different number of clusters. We observed that increasing the number of clusters improved the imputation of missing values (Figure {@fig:missingness}B-F). Additionally, we performed the same experiment by clustering the data with different weights. Interestingly, weight selection barely affected imputation performance, indicating that clustering peptides based on sequence only leads to the imputation of missing values as accuratley as when using the phosphorylation signal (Figure {@fig:missingness}F-I). These results indicate that DDMC clearly outperforms standard imputation strategies such as using 0s or the peptides' mean or minimum signal, and imputes missing values as accurately as matrix completion by PCA after optimizing hyperparameter selection.


### DDMC correctly predicts AKT1 and ERK2 as upstream kinases of signaling clusters containing their substrates

![**Validation of upstream kinase predictions.** (A-B) PCA analysis of the phosphorylation levels of MCF7 cells subjected to a drug screening [@PMID:31959955]. C) Upstream kinase prediction of EBDT's AKT1 and CPTAC's AKT1 cluster. D) Heatmap showing the effect of inhibitors to the phosphorylation signal of cluster 1. E) PhosphoSitePlus-reported AKT1 substrates per cluster (F) Resulting PSSM generated using ERK2 substrates reported by Carlson et al [@PMID:22028470]. (G) Upstream kinase predictions of CPTAC clusters 7, 9, 13, and 21 in addition to the ERK2 motif shown in (F). H) Upstream kinase predcitions of the same PSSMs after randomly shuffling the motif positions.](FigureM3.svg "Figure 3"){#fig:val width="100%"}

<!-- AKT cluster via MCF7 drug screening -->
DDMC is a tailored method that clusters mass spectrometry-generated phosphosites using its phosphorylation behavior and sequence information. A major benefit of modeling the sequence information is the construction of cluster motifs which can be useful to infer what putative upstream kinases might preferentially target peptides of a specific cluster. As shown later in this study (Figures. {@fig:TvsNAT}, {@fig:STK11}, {@fig:TI}), we utilized DDMC to analyze the phosphoproteomes of 110 treatment-naïve lung adenocarcinoma (LUAD) tumors and 101 paired normal adjacent tissues (NATs) from the National Cancer Institute (NCI)’s Clinical Proteomic Tumor Analysis Consortium (CPTAC) LUAD study [@PMID:32649874]. To validate the ability of our model to make upstream kinase predictions, we used DDMC to cluster the signaling status of MCF7 cells treated with a panel of 61 drug inhibitors reported by Hijazi et al [@PMID:31959955]. PCA analysis of the resulting cluster centers clearly identified the inverse correlation between the scores of AKT targeted inhibitors and the loading of cluster 1, indicating that the cluster's overall signal is downregulated by the presence of these compounds (Figures. {@fig:val}A-B). A heatmap displaying cluster's 1 phosphorylation signal across treatments corroborates that these peptides are dramatically downreagulated when treated with AKT/mTOR/PIK3 inhibitors (Figure. {@fig:val}C). Encouragingly, the specificity profile of AKT—within a collection of 55 different kinase profiles—most closely matches the PSSM of cluster 1 (Figure. {@fig:val}D). Moreover, after interrogating PhosphoSitePlus kinase-substrate interaction data set we observed that cluster 1 is indeed the cluster with the largest enrichment of AKT substrates (Figures. {@fig:val}F).

<!-- ERK2 substrates -->
Next, we extracted the sequences of ERK2 substrates identified in Carlson et al to create an ERK2-specific PSSM positive control (Figure. {@fig:val}F). As expected, ERK2 was predicted to be the upstream kinase with the highest preference for the cluster's motif (Figure. {@fig:val}G). Finally, given the consistent enrichment of hydrophobic and polar residues throughout the entire ERK2 target motif (Figure. {@fig:val}F), we asked whether randomly shuffling all cluster PSSM positions surrounding the phosphoacceptor residue would affect the upstream kinase prediction. Interestingly, this led to a 2-fold increase in the distance between ERK2 specificity profile and the cluster motif (Figures. {@fig:val}G and H). Furthermore, we subjected those clusters from the CPTAC data set that were preferentially favored by ERK2 to the same experiment. As expected, a similar magnitude of the decline in specificity. Note that the better prediction of ERK2 to the cluster of experimentally validated peptides (ERK2+) compared to CPTAC ERK2 motifs is not surprising given that whereas the former group contains only 26 peptides, the CPTAC clusters include hundreds of phosphosites. Overall, this experiment generally shows that despite the consistent biophysical properties of ERK2 target motif across positions the relative enrichment of hydrophobic and polar residues in each position determines the extent of specificity of ERK2 for a given motif (Figures. {@fig:val}G and H). Altogether, these results illustrate two different scenarios in which the cluster sequence information was successfully utilized to correctly identify upstream kinases modulating groups of phosphopeptides.

### A dual data-motif strategy improves prediction of different phenotypes and provides more robust clustering

![**Sequence information enhances model prediction.** A) Performance of a regression model predicting the mutational status of STK11 (blue) EGFR and/or ALK (yellow) and TI (green) in LUAD patients using either only phosphorylation data (weight=0), mainly sequences information (50), or a mix (15, 20, 40). B) MSE between the phosphorylation signal of 2000 randomly selected peptides and the center of its assigned cluster using a weight of 0 (data), 20 (mix), or 50 (sequence). C) Cumulative PSSM enrichment across positions comparing the data, mix, and sequence clustering strategies. (D-H) TBC1D5 S584-p peptide p-signal MSE (D), cumulative PSSM enrichment (E), and PSSM logo plots (F-H).](figureM4.svg "Figure 4"){#fig:preds width="100%"}

<!-- Intro to model set up and data -->
We utilized DDMC with the binomial sequence distance method and 24 clusters (Figure {@fig:method}, {@fig:missingness}B). We were able to include many peptides that were not observed in every tumor through our ability to handle missing data, but still filtered out any peptides that were not captured in at least two 10-plex TMT runs, resulting in a data set of 30,561 peptides. We used this fitting result throughout the rest of this study. The resulting 24 cluster motifs can be found in Figure {@fig:motifsS}.

<!-- Predictions -->
To evaluate the benefit of incorporating the peptide sequence information into the clustering criterion, we asked whether utilizing DDMC with different sequence weights would affect the performance of a regularized logistic regression model that predicts the mutational status of STK11, whether a patient harbors a mutation in EGFR and/or a gene fusion in ALK (EGFRm/ALKf), and the level of TI ("Hot" versus "Cold"). We found that for all three phenotypes, when the method only uses the phosphorylation signal (weight=0), the patient samples tend to be classified with less accuracy as when a combination of both data and sequence is used. Interestingly, in the case of STK11, the use of the largest weight wherein mainly the sequence motifs are used for clustering provided the best prediction performance. Likewise, EGFRm/ALKf samples were best classified with a mix weight of 15 or mainly prioritizing the sequence information. Finally, the regression model classifying whether a sample is "hot-tumor-enriched" (HTE) or "cold-tumor-enriched" (CTE) showed the best fitness with a weight of 40. Together, these results indicate that observing the motif information during clustering leads to final clusters that enhance the performance of regularized regression models predicting different phenotypes (Figures {@fig:preds} A and {@fig:predsS}).

<!-- Model behavior -->
Next, we explored how using different weights affects the overall phosphorylation signal and sequence information of the resulting clusters. To do so, we compared the model behavior after clustering the CPTAC data with a weight of 0 (data only), 20 (mix), and 50 (mainly sequence). First, we hypothesized that the data model would generate clusters with less within-cluster phosphorylation signal variation and thus a lower mean squared error (MSE). To test this, we computed the average peptide-to-cluster MSE of 2000 randomly selected peptides for each model. Although the differences were not significant, we did observe an direct correlation between weight and MSE (Figure {@fig:preds} B). Next, we calculated the cumulative PSSM enrichment by summing the information (bits) of all cluster PSSMs per model. Importantly, we observe a substantial increase in cumulative cluster information as the weight increases (Figure {@fig:preds} C). To further illustrate the clustering behavior, we tracked the phosphosite TBC1D5 S584-p in the three models. Consistent with the general trend, the Data and Mix models generated lower p-signal MSE when compared to its cluster signal than the Sequence model whereas weight correlates with the total PSSM enrichment (Figures {@fig:preds} D-E). Next, we quantified whether beyond absolute enrichment the Mix and Sequence models generated more representative PSSMs of TBC1D5 S584-p sequence than the Data model. To do so, we computed the mean of all pairwise PAM250 scores between the query sequence and all cluster sequences across models which clearly confirms that as the sequence prioritization of the model increases, the cluster PSSM is not only more enriched across all positions but also displays a more representative sequence of TBC1D5 phosphosite (Figures {@fig:preds} F-I).  Conclusively, these results show that using a mixed weight that similarly prioritizes both information sources—data and sequence—leads to more robust clustering of phosphosites displaying similar phosphorylation behavior and sequence motifs.  

### Widespread, dramatic signaling differences between tumor and normal adjascent tissue

![**Conserved tumor differences compared to normal adjascent tissue.** A) Hierarchical clustering of DDMC cluster centers. B–C) Principal components analysis scores (B) and loadings (C) of the samples and phosphopeptide clusters, respectively. D) Phosphorylation signal of tumor and NAT samples per cluster and statistical significance according to a Mann Whitney rank test. E) Receiver operating characteristic curve (ROC) of a logistic regression model. F) Logistic regression weights per cluster. G) Upstream kinase predictions of clusters 11 and 12. (H) NetworKIN kinase predictions of cluster 12 (I-J) Gene ontology analysis and (K-L) representative peptides of enriched biological processes of clusters 11 and 12.](figureM5.svg "Figure 5"){#fig:TvsNAT width="100%"}

<!-- The signaling in both tissues is so different that any modeling strategy captures distinct patterns -->
We explored whether DDMC could recognize conserved signaling patterns in tumors compared to normal adjacent tissue (NAT). The signaling difference between tumors and NAT samples was substantial, highlighting the significant signaling rewiring that tumor cells must undergo (Figure {@fig:TvsNAT}A). Using principal components analysis, we could observe that NAT samples were more similar to one another than to each tumor sample (Figure {@fig:TvsNAT}B/C). Nearly every cluster was significantly different in its average abundance between tumor and NAT (Figure {@fig:TvsNAT}D). Not surprisingly given these enormous differences, samples could be almost perfectly classified using their phosphopeptide signatures, with or without DDMC (Figure {@fig:TvsNAT}E; {@fig:TvsNATS}). Using the DDMC clusters, a logistic regression model identified that only cluster 11 contributes to the prediction of NAT samples (Figure {@fig:TvsNAT}C.

<!-- NEK1&2 and ciliagenesis -->
With the abundance changes and regression results we observed, we decided to further explore clusters 11 and 12. Cluster 11 shows a PSSM motif that might be preferred by NEK1, 2, and 4, and an enrichment of peptides involved in gas and oxygen transport, as well as several cytoskleleton remodeling or migration-related phenotypes according to a Gene Ontology (GO) analysis (Figures {@fig:TvsNAT}G and I). Even though NEKs are a largely understudied family of serine/threonine kinases, NEK1&2 have an established role in the formation and disassembly of cilia and NEK4 has also been implicated in regulating microtubule dynamics and stability [@PMID:22040655; @PMID:24921005]. The primary cilium serves as a signaling hub via the local expression of cell surface receptors and signaling molecules to sense environmental stimuli and thus regulate a handful of phenotypes including adaptation to hypoxia, migration, and escape from apoptosis [@PMID:30884815; @PMID:18381407]. Cancer cells typically lack cilia which could promote the emergence of these malignant phenotypes. Cluster 11 displays a striking phosphorylation decrease in tumor samples compared with NATs which could be representative of the presence or lack of NEK1&2 signaling, respectively. Within this group of peptides, there is a notable overrepresentation of hemoglobin subunits (HBG1, HBD, HBB, and HBA2) which could illustrate the different of oxygenation status of NATs versus malignant tissues as well as cytoskletal-remodeling proteins such as PEAK1, FLNA, GAS2L2, MARCKS, PEAK1, and ARHGEF7. All of these signaling molecules are downregulated in tumors whereas remain upregulated in NAT samples (Figure {@fig:TvsNAT}K).

<!-- CK2 and DNA damage response -->
On the other hand, CK2 favors the sequence motif of cluster 12 suggesting that this kinase might be a key modulator of this group of peptides. This association is also established by NetPhorest which underscores the presence of multiple experimentally validated CK2 substrates in this cluster (Figure {@fig:TvsNAT}J). GO analysis of cluster 12 determined a substantial enrichment in negative regulation of DNA duplex unwinding and pre-replicative complex assembly involved in cell cycle DNA replication. (Figures {@fig:TvsNAT}G, I-J). DNA duplex unwinding and replication are important procssess that play a major role in maintaining genome stability. DNA helicases are the enzymes responsible for unwinding the DNA and thus are essential for DNA replication. As such, they have been widely associated with DNA damage response (DDR) and cancer development [@PMID:23842644]. Interestingly, CK2 has been widely implicated in modulating DNA repair signaling pathways in response to DNA damage to promote cell survival in cancer [@PMID:25852190; @PMID:20164059; @PMID:27306791]. Finally, a study found that the CK2 inhibitor CX-4945 blocked DDR induced by gemcitabine and cisplatin and synergizes with these compounds in ovarian cancer cell lines [@PMID:22267551]. Cluster 12 contains several signaling proteins related to DNA replication and genome stability such as MCM3&4, the p53 interactor TP53BP1, BRCA1, ATRX, CENPF, and CDKs which are downregulating in NATs and upregulated in tumor samples. Conclusively, these results could suggest that CK2 might regulate signaling molecules within cluster 12 involved in DNA repair pathways to induce survival of cancer cells. Taken together, DDMC builds phosphoproteomic clusters that present signaling dysregulation common to tumors compared to NATs and identifies putative upstream kinases modulating them. These features can help to interpret phosphoproteomic results, even when there are substantial differences between samples.

### Genetic driver mutations are associated with more targeted phosphoproteomic rewiring

![**Phosphoproteomic aberrations associated with STK11 mutational status.** A) Phosphorylation signal of STK11 WT and mutant samples per cluster and statistical significance according to a Mann-Whitney rank test B) ROC of a logistic regression model predicting the STK11 mutational status and (C) its corresponding weights per sample type (D) Putative upstream kinases of clusters 7, and 8. (E) Representative cohesin loading peptides in cluster 7 (F-G) GO analysis and representative peptides of cluster 8.](figureM6.svg "Figure 6"){#fig:STK11 width="100%"}

<!-- STK11 DDMC data  -->
Inactivating somatic mutations in STK11 lead to increased tumorigenesis and metastasis [@PMID:17676035]. Thus, we aimed to identify the phosphoproteomic aberrations triggered by this genetic event. The majority of clusters were significantly altered, generally toward higher abundances with a mutation (Figure {@fig:STK11}A). The cluster centers corresponding to each patient's tumor and NAT samples could succesfully predict the STK11 mutational status by regularized logistic regression (Figure {@fig:STK11}B). The underlying signaling differences were mainly explained by the tumor samples of clusters 8 and 14 to which the model assigned very large negative weights (Figure {@fig:STK11}C). These results motivated further exploration of clusters 7, 8 and 14 which present sequence motifs favored by ERK2, CK1/BRCA1/PKD, and PKA/PKC/PKD, respectively (Figure {@fig:STK11}D).

<!-- ERK2 cohesin loading -->
Cluster 7 is highly enriched with peptides involved in regulation of the cell cycle  by cohesin loading. Cohesin is a protein complex that mediates sister chromatid cohesion by directly binding with DNA. This interaction holds both chromatids together after DNA replication until anaphase wherein cohesin is removed to facilitate chromosome segregation during cell division. Cluster 7 contains the inhibitor phosphosite of the tumor suppressor RB1 S795-p, the member of the cohesin loading complex NIPBL (S280-p, S280-p;S284-p, and S350-p), and the cohesin release factor WAPL (S221-p and S221-p;S223-p). Studies have shown that RB1 inactivation can lead to defects in chromosome cohesion that in turn compromises chromosome stability [@PMID:20551165; @PMID:24613344]. Manning et al demonstrated that depletion of WAPL in RB1-deficient cells promoted cohesin association with chromatin [@PMID:24613344]. Among these phosphosites, we observed strong opposing signals between STK11 WT and mutant patients in NIPBL S280-p, WAPL S221-p;S223-p and RB1 S795-p (Figure {@fig:STK11}E) which reinforces the link between STK11 activity and chromatin instability. Moreover, CDCA5 is key regulator of sister chromatid cohesion by stabilizing cohesin complex association with chromatin and it was identified as a prognostic factor of lung cancer through a tumor tissue microarray analysis of 262 non–small cell lung cancer (NSCLC) patients [@PMID:20551060]. Interestingly, they showed that CDCA5 phosphorylation of S209 by ERK2 enhanced cell proliferation [@PMID:20551060]. Therefore, these results might suggest that mutations inactivating mutations in STK11 might correlate with signaling defects in sister chromatid cohesion during the cell cycle which in turn could lead to chromosome instability and cell cancer growth. 

<!-- CK1 and Golgi coating -->
The signal of phosphosites of cluster 8 specifically in tumor samples exaplains the signaling differences between STK11 WT and mutant samples (Figure {@fig:STK11}C). This cluster presents a clear enrichment of peptides involved in the regulation of the Golgi apparatus such as GOLGA2-5, GOLGB1, and GOLPH3. Cancer cells commonly undergo fragmentation of the Golgi which has been shown to drive several malignant molecular signatures including the hyperactivity of motor proteins and kinase signaling dysregulation [@PMID:27064441]. Myosin 18A and 1E pertain to cluster 18 and the former has been reported to interact with GOLPH3 to induce Golgi dispersal. Moreover, a series of studies uncovered that GOLPH3 promotes cell proliferation in cancer [@PMID:23006319; @PMID:22675169; @PMID:23132295]. The phosphorylation behavior of GOLPH3, Myosin 18A, and GOLGA2 in STK11 WT compared with STK11 mutant patients shows a dramatic increase of abundance in the latter which supports the association between STK11 activity and an oncogenic role of the Golgi apparatus in these patients. Protein kinases pertaining to this cluster include BCR, EGFR, SLK, and mTOR—as well as its regulators LAMTOR and RICTOR—whose connection to Golgi fragmentation has not been made hitherto. However, there was no meaningful signaling difference between STK11 WT and mutant specimens in any of them. Together, these results suggest that STK11 mutations in tumor samples could affect the dispersion of the Golgi apparatus compared with STK11 WT samples.

<!-- EGFRm/ALKf -->
Tyrosine kinase inihibitors (TKIs) targeting the receptor tyrosine kinases (RTKs) EGFR and ALK are effective treatments in cancer patients with EGFR mutations and/or ALK translocations (EGFRm/ALKf). However, these treatments are limited by drug resistance which in some cases can be mediated by the concomitant signaling of both RTKs activated by driver mutations [@PMID:21791641; @PMID:27707887]. Once again, the signaling cluster centers allowed a regularized logistic regression model to accurately classify samples according to its EGFRm/ALKf status (Figure {@fig:EGFRm/ALKf}). 

<!-- Comparison to other regression strategies -->
Finally, we compared the classification performance of four regularized logistic regression models fit to either the DDMC clusters, clusters generated by the standard methods GMM and k-means, or the raw phosphoproteomic data directly. It is worth noting that unlike DDMC, methods such as GMM, k-means, or direct regression can not handle missing values and thus for this strategies we used the 1,311 peptides that were observed in all samples, whereas DDMC was fit to the entire data set comprising 30,561 phosphosites. Interestingly, we found that samples were classified with higher accuracy with DDMC than using GMM and similarly than with k-means, especially with STK11 (Figure {@fig:RegressionsS}A). Direct regression to the raw signaling data yielded excellent performance, however this strategy assigns thousands of coefficients to different peptides that vary every time the model is run which renders this appraoch unable to establish a consistent link between mutations and signaling (Figure {@fig:RegressionsS}). This analysis illustrates the association between STK11 activity with two novel phenotypes, namely chromosome cohesion during cell cylce and Golgi fragmentation, and proposes putative mechanisms to support it.

### Exploration of immune infiltration-associated signaling patterns in tumors   

![**Phosphoproteomic signatures driving tumor immune infiltration.** (A) Phosphorylation abundance of CTE and HTE samples per cluster and statistical significance according to a Mann-Whitney rank test (B-C) ROC and coefficients of a logistic regression model predicting infiltration status—cold-tumor enriched (CTE) vs hot-tumor enriched (HTE)—(D) Putative upstream kinases of clusters 2 and 6 and (E-F) its GO enrichment analysis (G-H) ROC and coefficients of a logistic regression model predicting the TI with only clusters 21 and 24 (I) Upstream kinases of clusters 21 and 24 and (J-K) GO analysis.](figureM7.svg "Figure 7"){#fig:TI width="100%"}

<!-- Intro to immunotherapy in NSCLC -->
Immune checkpoint inhibitors (ICIs) have emerged as effective treatment options for NSCLC patients. However, there still is a need to identify or influence which patients will respond to these therapies. Patients that do not respond to ICIs have tumors that present immune-cold behavior either inherently, or via an adaptive process after long exposure to the drug. However, the signaling mechanism by which malignant cells prevent TI (TI) remains elusive. To this end, we used our DDMC clusters to explore the shared signaling patterns that differentiate "hot-tumor-enriched" (HTE) from "cold-tumor-enriched" CTE LUAD patients [@PMID:32649874; @PMID:29141660]. HTE and CTE status per patient was determined using xCell by Gilette et al[@PMID:32649874].

<!-- Data -->
We observed that four clusters were significantly different in their average abundance between HTE and CTE samples (Figure {@fig:TI}A). Cluster 17, 18, and 20 display significantly higher abundances in HTE compared to CTE samples whereas cluster 21 presents the opposite trend. Samples could be accurately classified using the DDMC clusters (Figure {@fig:TI}B). This predictive performance was mainly explained by large contribution of cluster 2 to predict HTE samples as well as that of cluster 6 to predict CTE specimens. Additional clusters such as 9, 10, 15, 17, 20, and 21 contributed to explain the signaling differences between both groups but to a lesser extent (Figure {@fig:TI}C). 

<!-- Interpretation -->
These results prompted us to further investigate clusters 6, 17, 20, and 21 which our model predicts to be regulated by CK1&PKA, STK11&p38, CK2&STK11, and ERK2, respectively (Figure {@fig:TI}D). When exploring immunologically relevant phenotypes in the GO analysis of each cluster, we observed that clusters 6, 17, and 20 showed a substantial over-representation of immunological processes. Conversely, neither of these phenotypes were present in the GO analyses of cluster 2 nor cluster 21 wherein the former substantially contributes to predict HTE samples and the latter shows a significant increase of phosphorylation abundance in CTE over HTE samples {@fig:TI}A, C). Cluster 6 members are involved in mediating B cell homeostasis, T cell differentiation, T cell receptor signaling, and regulation of T cell activation. In addition, this cluster presents a slight under-representation of overall, adaptive, and humoral immune responses (Figure {@fig:TI}E). Cluster 17 presents an over-representation of several innate and adaptive immune response pathways whereas cluster 20 is enriched by responses orchestrated by the innate immune system (Figure {@fig:TI}F-G). The transcription factor NFATC crucially regulates T cell activation and proliferation and several studies show that the predicted upstream kinase of cluster 20 CK2 directly phosphorylates this protein and regulates its gene expression [@PMID:10652349; @PMID:31988467]. In addition, CK2 has also been shown to phosphorylate Regulators of Calcineurin (RCAN) proteins, which indirectly regulate NFATC function [@PMID:23732701]. Multiple peptides of both NFATC and RCAN are members of cluster 20 which reinforces the role of CK2 in promoting immune infiltration in lung cancer patients. Intriguinly, inactivating mutations in STK11 have been reported to promote anti-PD1/PD-L1 resistance in KRAS-mutant LUAD suggesting a key role of STK11 in modulating tumor immune infiltration [@PMID:29773717]. All in all, these data demonstrate that the presence or lack of tumor immune infiltration can be accurately predicted by the DDMC clusters which in turn help identify putative upstream kinases modulating immune evasion.  
