## Discussion

<!-- Paper Summary / Intro -->
Phosphorylation-based cell signaling through the coordinated activity of protein kinases enables cells to swiftly integrate environmental cues and orchestrate a myriad of biological processes. MS-based global phosphoproteomic data provides the unique opportunity to globally interrogate signaling networks to better understand cellular decision-making and its therapeutic implications. However, these data also present challenging issues as a consequence of their incomplete and stochastic coverage, high-content but low-sample throughput, and variation in coverage across experiments. Here, we propose a clustering method, Dual Data and Motif Clustering (DDMC), that untangles highly complex coordinated signaling changes by grouping phosphopeptides based on their phosphorylation behavior and sequence similarity (Figure {@fig:method}). To test the utility of DDMC, we clustered the phosphoproteomes of LUAD patients and used the resulting groups of peptides to decipher signaling dysregulation common to tumors, genetic backgrounds, and tumor infiltration status (Figures {@fig:TvsNAT}, {@fig:STK11}, {@fig:TI}).

<!-- Sequence information improves prediction and interpretation -->
Previous efforts in regressing mass spectrometry-based phosphorylation measurements against phenotypic or clinical data have been based on the ability of certain regression models such as PLSR or LASSO to robustly predict using high-dimensional and correlated data [@PMID:25750696]. While these models can generally be predictive with such data, they are not easily interpretable (Figure {@fig:TvsNATS}B). Hence, we hypothesized that clustering large-scale MS measurements based on biologically meaningful features and utilizing the cluster centers to fit regression methods could enhance the predictive performance of the model while providing highly interpretable results wherein clusters constitute signaling nodes distinctly correlated with cell patient phenotypes. Here, we demonstrate that DDMC enhances model prediction and interpretation (Figures {@fig:preds}A, {@fig:RegressionsS}, {@fig:val}).

<!-- Caveats of upstream predictions -->
Model interpretation is enhanced by comparing the resulting cluster PSSMs with kinase specificity data such as PSPL to identify putative upstream kinases modulating signaling clusters. Computational validations showed that DDMC was able to correctly associate AKT1 and ERK2 with clusters of their respective substrates (Figure {@fig:val}). It is worth noting, however, that kinase specificity is defined by additional features beyond the phosphosite motif such as kinase-substrate co-localization, regulation by phosphosite-binding domains (e.g., SH2, PTB domains), or docking. In addition, a major limitation of PSPL experiments is that since they do not provide docking information, the real affinity between the string of identified peptide residues as key determinants of specificity of a sequence motif and the interacting kinase domain is unknown. This limitation could also compromise kinase-cluster associations established by DDMC. A method combining bacterial surface-display of peptide libraries with next-generation sequencing tackles this limitation by quantifying the specificity of a kinase to virtually all possible motif combinations [@PMID:29547119]. Thus, as the number of profiled kinases with this technique increases, these measurements could be used to rank cluster peptides by magnitude of specificity to a specific kinase to make better upstream kinase predictions.

<!-- Data-driven strategy -->
A key benefit of DDMC is that the identified clusters are not limited to pre-existing motifs and are therefore not dependent on prior experimentally validated kinase-substrate interactions. Thereby, this method could improve our understanding of the signaling effects of understudied kinases. For instance, our model predicts NEK1&2 promote, at least in part, a cluster with strikingly increased signaling in NATs compared to tumors. Further exploration of this cluster led us to hypothesize that the lack of NEK signaling in tumor samples might associated with the absence of ciliagenesis and adaptation to hypoxia in lung tumors {@fig:TvsNAT}G-H). Additionally, we show that cluster 8, which greatly contributes to explain the signaling differences between STK11 WT and mutant samples in tumors (Figure {@fig:STK11}C), is enriched with proteins such as GOLPH3 and Myosin 18A that have been shown to promote Golgi fragmentation in cancer [@PMID:23006319; @PMID:22675169; @PMID:23132295]. This prompts us to consider the novel interaction between CK1 and these signaling molecules.

<!-- Modeling missingness -->
An additional major challenge being faced during the analysis of large-scale signaling data is missingness. Given that statistical tools often require complete data sets, researchers use standard methods to impute missing values such as the peptides’ mean or minimum signal, constant zero, or PCA imputation only in peptides wherein at least 50% of their samples were required to have non-missing values as excessive missing values can result in poor imputation [@PMID:32649874; @PMID:32649875; @PMID:32033228]. In this study we show that DDMC can model a data set of 30,561 peptides after filtering out any phosphosites that were not captured in at least 2 TMT (up to ~80% of missingness) by ignoring unobserved values during EM distribution estimation and calculation of GMM probabilities (see methods). Therefore, this method enables clustering of signaling data despite a remarkable number of missing values. Furthermore, DDMC clearly outperforms the imputation performance of using the peptides’ mean, minimum signal, or constant zero and provides similar results to PCA imputation. This important feature could offer the possibility of conducting pan-cancer phosphoproteomics studies using readily available large-scale clinical phosphoproteomic data.

<!-- Models combining info in models -->
The benefit of building algorithms combining different information sources to build better models is also shown in previously published approaches. For instance, INKA predicts active kinases by integrating two scores illustrating the a kinase's activation status accordingo to its phosphorylation status, and two scores that quantify the abundance of its substrates [@PMID:30979792]. In another study, Exarchos et al. formulated a decision support system that integrates clinical, imaging, and genomic data to identify the factors that contribute to oral cancer progression and predict relapses. The authors found that combining the more accurate individual predictors yielded better predictions than those generated by other strategies reported in the literature [@PMID:21859630]. Additionally, BOADICEA is a method that allows systematic risk stratification of breast cancer patients by incorporating the effects of lifestyle, hormonal and reproducrive risk factors, mammographic density, and of the common breast cancer susceptibility genetic variants into the prediction model.

The benefit of building algorithms combining different information sources is evident in previously published approaches. For instance, INKA predicts active kinases by integrating scores reflecting both phosphorylation status and substrate abundance [@PMID:30979792]. In another study, Exarchos et al. formulated a decision support system that integrates clinical, imaging, and genomic data to identify the factors that contribute to oral cancer progression and predict relapses. The authors found that combining the more accurate individual predictors yielded better predictions than those generated by other strategies reported in the literature [@PMID:21859630]. Finally, BOADICEA is a method that allows systematic risk stratification of breast cancer patients by incorporating the effects of lifestyle, hormonal and reproductive risk factors, mammographic density, and of the common breast cancer susceptibility genetic variants into the prediction model [@PMID:18349832].

<!-- Conclusion / Closure -->
In total, in this study we show that combining the information about the sequence features and phosphorylation abundance leads to more robust clustering of global signaling measurements. Use of the DDMC clusters to regress against cell phenotypes led to enhanced model predictions and interpretation. Thus, we propose DDMC as a general and flexible strategy for phosphoproteomic analysis.
