## Supplementary Figures {.page_break_before}

### Sequence Modeling

DDMC simultaneously models phosphorylation signal and peptide sequence to cluster global signaling measurements. The user-defined variables of this algorithm are the number of clusters and the scaling factor or weight to prioritize the data or the sequence scores. During the Expectation– or E-step, the algorithm guesses a data and a sequence probability distribution given the current model parameters to assign every peptide into a cluster based on its maximum likelihood estimation. That is, the GMM generates the probabilities of each phosphosite being assigned to every cluster according to its log-transformed abundance. In the PAM250 method, the average distance between a peptide’s sequence and each of those within a cluster are calculated using the transition matrix. For the binomial method, we collected 333,492 random MS-generated phosphopeptides from the public compendium PhosphoSitePlus13 to be used as the set of background sequences. The probability of observing each amino acid at each position was calculated by comparing each cluster and the background using an incomplete beta function. First, the cluster sequences are used to build counts matrices displaying the number of amino acid occurrences across positions. These matrices are then individually compared to the weighted frequencies matrix of the background set of sequences using a binomially distributed model in order to generate a binomial probability matrix for each cluster. The probability of a peptide being assigned to any cluster is the average of each position’s probability present in the correspondinog cluster’s binomial probability matrix. For each cluster, the final score is the sum of the GMM and sequence scores multiplied by a user-defined weight which scales both scores so that a weight of 0 or a very large number only scores the data or sequence, respectively. Importantly, the final scores represent the relative probabilities that define the partial membership of the query peptide to each cluster. In other words, the EM algorithm provides soft cluster assignments for every peptide. Then, the GMM parameters and cluster sequences are updated during the M-step based on the assignments generated during the previous E-step. These two steps repeat until convergence.

![**Schematic of the DDMC simultaneous data and peptide sequence clustering approach.** Peptides are initialized into clusters at random. The measurements of abundance are represented by a multivariate Gaussian mixture model where each dimension of the distribution represents the abundance within a sample. Next, an expectation-maximization fitting scheme is used. During the expectation step, the distance of each peptide sequence to each cluster is calculated. This is done either through a binomial enrichment scheme (method 1) or using the average PAM250 distance (method 2). In parallel, the distance of each peptide abundance is compared to the cluster centers. These two distances are combined to update the assignments of each peptide to each cluster. During the maximization step, the cluster centers of the data are updated based on the weighted average of the peptide abundances in each condition. The peptide motifs are similarly updated through a weighted combination of the assigned peptides. Both steps continue sequentially until the change in peptide assignments between each iteration drops below a threshold.](EM_diagram.svg){#fig:methodS width="100%", tag="S1"}


### Logo plots of all Cluster Motifs {.page_break_before}

![**Logo plots of all Cluster PSSMs.** (A-U) Sequence motifs of clusters 1 through 24.](figureMS2.svg){#fig:motifsS tag="S2"}

### A dual data-motif strategy improves prediction of different phenotypes
![**Sequence information enhances model prediction** (A-O) Performance of a regression model predicting the mutational status of STK11 (A-E) EGFR and/or ALK (F-J) and tumor infiltration level (hot versus cold) (K-O) in LUAD patients using either only phosphorylation data (0), mainly peptide sequences (50), or a mix (15, 20, 40). EGFRm/ALKf; EGFR mutant and/or ALK fusion.](figureMS3.svg){#fig:predsS tag="S3"}

### Implementing different modeling strategies to find tumor differences compared to NATs {.page_break_before}

![**Additional modeling strategies to find conserved tumor differences compared to NATs** A) ROC plot of a logistic regression model fit to the complete portion of the phosphoproteomic CPTAC LUAD data set. B) Phospho-peptides with largest weights explaining the observed differences between tumors and NATs. C-D) ROC plot of a logistic regression model fit to the complete portion of the signaling data set clustered by k-means (D) and its corresponding cluester coefficients (D). E) p-site abundance between NAT and tumor patients per k-means cluster and its statistical significance. (F-H) Analysis using GMM clustering.](figureMS4.svg){#fig:TvsNATS tag="S4"}

### Regressing DDMC clusters against the mutational status of EGFR/ALK {.page_break_before}

![**Prediction of patient samples harboring EGFR mutations or ALK fusions (EGFRm/ALKf)** A) Phosphorylation signal of DDMC clusters grouped by EGFRm/ALKf mutantants and WT samples. Its statistical significance is indicated on the top part of the plot via a series of Mann Whitney rank test. B) ROC plot of the logistic regression model fit to the DDMC clusters. (D-G) Logo plots of the sequence motifs for clusters 2, 13, 20, and 23. H) Prediction of upstream kinases corresponding to the aforementioned clusters.](figureMS5.svg){#fig:EGFRm/ALKf tag="S5"}

### Comparing the predictive of performance of using DDMC to different regression strategies {.page_break_before}

![**Comparing the predictive performance of a logistic regression model using different modeling strategies.** (A-D) Performance of a logistic regression model predicting STK11 mutational status using DDMC clusters (A), unclustered signaling data (B), k-means clusters (C) and GMM clusters (D). Predictive performance of a logistic regression model regressed against EGFRm/ALKf mutational status. Note that a complete portion of the entire data set containing 1311 peptides was used for every modeling strategy other than DDMC. DDMC was fit to 30561 peptides that include at least a minimum of 2 10-plex TMT experiments.](figureMS6.svg){#fig:RegressionsS tag="S6"}
