## Supplementary Figures

<!-- (Supp. Figure 1) -->
### Sequence Modeling

DDMC simultaneously models phosphorylation signal and peptide siequence to cluster global signaling measurements. The user-defined variables of this algorithm are the number of clusters and the scaling factor or weight to prioritize the data or the sequence scores. During the Expectation– or E-step, the algorithm guesses a data and a sequence probability distribution given the current model parameters to assign every peptide into a cluster based on its maximum likelihood estimation. That is, the GMM generates the probabilities of each phosphosite being assigned to every cluster according to its log-transformed abundance. In the PAM250 method, the average distance between a peptide’s sequence and each of those within a cluster are calculated using the transition matrix. For the binomial method, we collected 333,492 random MS-generated phosphopeptides from the public compendium PhosphoSitePlus13 to be used as the set of background sequences. The probability of observing each amino acid at each position was calculated by comparing each cluster and the background using an incomplete beta function. First, the cluster sequences are used to build counts matrices displaying the number of amino acid occurrences across positions. These matrices are then individually compared to the weighted frequencies matrix of the background set of sequences using a binomially distributed model in order to generate a binomial probability matrix for each cluster. The probability of a peptide being assigned to any cluster is the average of each position’s probability present in the correspondinog cluster’s binomial probability matrix. For each cluster, the final score is the sum of the GMM and sequence scores multiplied by a user-defined weight which scales both scores so that a weight of 0 or a very large number only scores the data or sequence, respectively. Importantly, the final scores represent the relative probabilities that define the partial membership of the query peptide to each cluster. In other words, the EM algorithm provides soft cluster assignments for every peptide. Then, the GMM parameters and cluster sequences are updated during the M-step based on the assignments generated during the previous E-step. These two steps repeat until convergence.

![**Schematic of the DDMC simultaneous data and peptide sequence clustering approach.** Peptides are initialized into clusters at random. The measurements of abundance are represented by a multivariate Gaussian mixture model where each dimension of the distribution represents the abundance within a sample. Next, an expectation-maximization fitting scheme is used. During the expectation step, the distance of each peptide sequence to each cluster is calculated. This is done either through a binomial enrichment scheme (method 1) or using the average PAM250 distance (method 2). In parallel, the distance of each peptide abundance is compared to the cluster centers. These two distances are combined to update the assignments of each peptide to each cluster. During the maximization step, the cluster centers of the data are updated based on the weighted average of the peptide abundances in each condition. The peptide motifs are similarly updated through a weighted combination of the assigned peptides. Both steps continue sequentially until the change in peptide assignments between each iteration drops below a threshold.](EM_diagram.svg){#fig:method width="100%", tag="S1"}


<!-- (Supp. Figure 2) -->
### Logo plots of all Cluster Motifs

![**Logo plots of all Cluster PSSMs.** (A-U) Sequence motifs of clusters 1 through 24.](figureMS1.svg){#fig:uncenSingle tag="S2"}

\pagebreak

<!-- (Supp. Figure 3) -->
### Implementing different modeling strategies to find tumor differences compared to NATs

![**Using other modeling strategies to find conserved tumor differences compared to NATs.** A) ROC plot of an L1 logistic regression model fit to the complete portion of the phosphoproteomic data set. B) Coefficients of phospho-peptides explaining the observed differences between tumors and NATs. C-D) ROC plot of an L1 logistic regression model fit to the complete portion of the signaling data set clustered by k-means (D) and its corresponding cluester coefficients (D). E) p-site abundance between NAT and tumor patients per k-means cluster and its statistical significance. F-G) ROC plot of an L1 logistic regression model fit to the complete portion of the signaling data set clustered by GMM (F) and its correspodning cluster coefficients (G). (H) p-site abundance between NAT and tumor samples per GMM cluster and its statistical significance.](figureMS3.svg){#fig:uncenSingle tag="S3"}

\pagebreak

<!-- (Supp. Figure 4) -->
### Implementing different modeling strategies to find signaling differences between TP53 WT and mutant patients

![**Comparing the signaling differences between TP53 mutant and WT patients found by using raw phosphoproteomic data, k-means clusters, or GMM clusters.** (A) ROC plot of an L1 logistic regression model fit to the complete portion of the phosphoproteomic data set. B) Coefficients of peptides explaining the observed differences between TP53 WT and mutant patients. (C-D) ROC plot (C) and cluster coefficients (D) of an L1 logistic regression model fit to the complete portion of the data clustered by k-means. (E) p-site abundance between TP53 WT and mutant patients and its statistical significance. (F-H) Re-implmention of plots C-E but clustering the complete portion with a GMM.](figureMS4.svg){#fig:uncenSingle tag="S4"}


\pagebreak

<!-- (Supp. Figure 5) -->
### Regressing DDMC clusters against the mutational status of EGFR/ALK and STK11

![**Prediction of EGFRm/ALKf and STK11 mutants.** A) TP53 cluster coefficients of an L1 logistic regression model. B) Phosphorylation levels of EGFR/ALK WT versus mutant patients and its statistical significance. C) ROC plot of an L1 logistic regression model fit to the DDMC clusters D) EGFRm/ALKf DDMC cluster coefficients of an L1 logistic regression model. (E-G) Re-implementation of plots B-D but predicting STK11 mutational status.](figureMS5.svg){#fig:uncenSingle tag="S5"}